<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Alejandro Ochoa" />

<meta name="date" content="2021-07-26" />

<title>Read and write BED files with genio</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>
<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Read and write BED files with <code>genio</code></h1>
<h4 class="author">Alejandro Ochoa</h4>
<h4 class="date">2021-07-26</h4>



<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>The <code>genio</code> (GenIO = Genetics I/O) package aims to facilitate reading and writing genetics data. The focus of this vignette is processing Plink BED/BIM/FAM files.</p>
<p>There are some limited alternatives for reading and/or writing BED files in R, which are slower and harder to use, which motivated me to write this package. Here we make direct comparisons to those packages, to illustrate the advantages of <code>genio</code>.</p>
</div>
<div id="generate-a-large-matrix-of-random-genotypes" class="section level1">
<h1>Generate a large matrix of random genotypes</h1>
<p>Let’s begin by creating a large genotype matrix with completely random data, and with missing values so this becomes non-trivial to read and write.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a><span class="co"># Data dimensions.</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true"></a><span class="co"># Choose non-multiples of 4 to test edge cases of BED parsers.</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true"></a><span class="co"># Number of loci.</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true"></a>m_loci &lt;-<span class="st"> </span><span class="dv">10001</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true"></a><span class="co"># Number of individuals</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true"></a>n_ind &lt;-<span class="st"> </span><span class="dv">1001</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true"></a><span class="co"># Overall allele frequency</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true"></a><span class="co"># (we don&#39;t do any actual genetics in this vignette,</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true"></a><span class="co"># so just set to a reasonable value)</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true"></a>p &lt;-<span class="st"> </span><span class="fl">0.5</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true"></a><span class="co"># Missingness rate</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true"></a>miss &lt;-<span class="st"> </span><span class="fl">0.1</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true"></a></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true"></a><span class="co"># Total number of genotypes</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true"></a>n_data &lt;-<span class="st"> </span>n_ind <span class="op">*</span><span class="st"> </span>m_loci</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true"></a><span class="co"># Draw random genotypes from Binomial</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true"></a>X &lt;-<span class="st"> </span><span class="kw">rbinom</span>( n_data, <span class="dv">2</span>, p)</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true"></a><span class="co"># Add missing values</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true"></a>X[ <span class="kw">sample</span>(n_data, n_data <span class="op">*</span><span class="st"> </span>miss) ] &lt;-<span class="st"> </span><span class="ot">NA</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true"></a><span class="co"># Turn into matrix</span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true"></a>X &lt;-<span class="st"> </span><span class="kw">matrix</span>(X, <span class="dt">nrow =</span> m_loci, <span class="dt">ncol =</span> n_ind)</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true"></a></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true"></a><span class="co"># Inspect the first 10 individuals at the first 10 loci</span></span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true"></a>X[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true"></a><span class="co">#&gt;       [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]</span></span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true"></a><span class="co">#&gt;  [1,]    0    2    2    2    2    1    0    2    1     1</span></span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true"></a><span class="co">#&gt;  [2,]    1    1    1    0    1    2    0    1    1     1</span></span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true"></a><span class="co">#&gt;  [3,]    1    0    0    2    1    0    1    0    2     1</span></span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true"></a><span class="co">#&gt;  [4,]    1    2    1    0    2    0    0    1   NA     1</span></span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true"></a><span class="co">#&gt;  [5,]    1    1    1    1   NA    1    1    2    1     0</span></span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true"></a><span class="co">#&gt;  [6,]   NA    1    0   NA    1   NA    2    2    2     1</span></span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true"></a><span class="co">#&gt;  [7,]    1    2    2    1    1    1    1    2    1     2</span></span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true"></a><span class="co">#&gt;  [8,]    1   NA    1    0    1    2    2    1    2     2</span></span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true"></a><span class="co">#&gt;  [9,]   NA   NA    0    0    2    1    1    0    0     1</span></span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true"></a><span class="co">#&gt; [10,]    1    1    2    2    1    0    1    1    2    NA</span></span></code></pre></div>
<p>To create annotation tables that look a bit more interesting than the defaults, let us create some slightly more realistic values. Here we use our first <code>genio</code> function!</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a><span class="kw">library</span>(genio)</span></code></pre></div>
<p>First we create and edit the locus annotations table.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a><span class="co"># We have to specify the number of loci</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true"></a>bim &lt;-<span class="st"> </span><span class="kw">make_bim</span>( <span class="dt">n =</span> m_loci )</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true"></a><span class="co"># Inspect the default values</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true"></a>bim</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true"></a><span class="co">#&gt; # A tibble: 10,001 × 6</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true"></a><span class="co">#&gt;      chr    id  posg   pos   ref   alt</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true"></a><span class="co">#&gt;    &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true"></a><span class="co">#&gt;  1     1     1     0     1     1     2</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true"></a><span class="co">#&gt;  2     1     2     0     2     1     2</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true"></a><span class="co">#&gt;  3     1     3     0     3     1     2</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true"></a><span class="co">#&gt;  4     1     4     0     4     1     2</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true"></a><span class="co">#&gt;  5     1     5     0     5     1     2</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true"></a><span class="co">#&gt;  6     1     6     0     6     1     2</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true"></a><span class="co">#&gt;  7     1     7     0     7     1     2</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true"></a><span class="co">#&gt;  8     1     8     0     8     1     2</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true"></a><span class="co">#&gt;  9     1     9     0     9     1     2</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true"></a><span class="co">#&gt; 10     1    10     0    10     1     2</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true"></a><span class="co">#&gt; # … with 9,991 more rows</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true"></a></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true"></a><span class="co"># Let&#39;s add the &quot;chr&quot; prefix to the chromosome values,</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true"></a><span class="co"># so we recognize them when we see them later.</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true"></a>bim<span class="op">$</span>chr &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&#39;chr&#39;</span>, bim<span class="op">$</span>chr)</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true"></a><span class="co"># Make SNP IDs look like &quot;rs&quot; IDs</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true"></a>bim<span class="op">$</span>id &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&#39;rs&#39;</span>, bim<span class="op">$</span>id)</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true"></a><span class="co"># Make positions 1000 bigger</span></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true"></a>bim<span class="op">$</span>pos &lt;-<span class="st"> </span>bim<span class="op">$</span>pos <span class="op">*</span><span class="st"> </span><span class="dv">1000</span></span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true"></a><span class="co"># Select randomly between Cs and Gs for the reference alleles</span></span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true"></a>bim<span class="op">$</span>ref &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">&#39;C&#39;</span>, <span class="st">&#39;G&#39;</span>), m_loci, <span class="dt">replace =</span> <span class="ot">TRUE</span>)</span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true"></a><span class="co"># Select randomly between As and Ts for the alternative alleles</span></span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true"></a>bim<span class="op">$</span>alt &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="st">&#39;A&#39;</span>, <span class="st">&#39;T&#39;</span>), m_loci, <span class="dt">replace =</span> <span class="ot">TRUE</span>)</span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true"></a></span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true"></a><span class="co"># Inspect the table with our changes</span></span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true"></a>bim</span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true"></a><span class="co">#&gt; # A tibble: 10,001 × 6</span></span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true"></a><span class="co">#&gt;    chr   id     posg   pos ref   alt  </span></span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true"></a><span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;</span></span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true"></a><span class="co">#&gt;  1 chr1  rs1       0  1000 C     A    </span></span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true"></a><span class="co">#&gt;  2 chr1  rs2       0  2000 G     T    </span></span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true"></a><span class="co">#&gt;  3 chr1  rs3       0  3000 C     T    </span></span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true"></a><span class="co">#&gt;  4 chr1  rs4       0  4000 G     A    </span></span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true"></a><span class="co">#&gt;  5 chr1  rs5       0  5000 G     T    </span></span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true"></a><span class="co">#&gt;  6 chr1  rs6       0  6000 G     A    </span></span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true"></a><span class="co">#&gt;  7 chr1  rs7       0  7000 C     A    </span></span>
<span id="cb3-45"><a href="#cb3-45" aria-hidden="true"></a><span class="co">#&gt;  8 chr1  rs8       0  8000 G     A    </span></span>
<span id="cb3-46"><a href="#cb3-46" aria-hidden="true"></a><span class="co">#&gt;  9 chr1  rs9       0  9000 G     T    </span></span>
<span id="cb3-47"><a href="#cb3-47" aria-hidden="true"></a><span class="co">#&gt; 10 chr1  rs10      0 10000 C     T    </span></span>
<span id="cb3-48"><a href="#cb3-48" aria-hidden="true"></a><span class="co">#&gt; # … with 9,991 more rows</span></span></code></pre></div>
<p>Now we similarly create and edit the individual annotations table.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true"></a><span class="co"># Specify the number of individuals</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true"></a>fam &lt;-<span class="st"> </span><span class="kw">make_fam</span>( <span class="dt">n =</span> n_ind )</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true"></a><span class="co"># Inspect the default values</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true"></a>fam</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true"></a><span class="co">#&gt; # A tibble: 1,001 × 6</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true"></a><span class="co">#&gt;      fam    id   pat   mat   sex pheno</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true"></a><span class="co">#&gt;    &lt;int&gt; &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true"></a><span class="co">#&gt;  1     1     1     0     0     0     0</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true"></a><span class="co">#&gt;  2     2     2     0     0     0     0</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true"></a><span class="co">#&gt;  3     3     3     0     0     0     0</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true"></a><span class="co">#&gt;  4     4     4     0     0     0     0</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true"></a><span class="co">#&gt;  5     5     5     0     0     0     0</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true"></a><span class="co">#&gt;  6     6     6     0     0     0     0</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true"></a><span class="co">#&gt;  7     7     7     0     0     0     0</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true"></a><span class="co">#&gt;  8     8     8     0     0     0     0</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true"></a><span class="co">#&gt;  9     9     9     0     0     0     0</span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true"></a><span class="co">#&gt; 10    10    10     0     0     0     0</span></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true"></a><span class="co">#&gt; # … with 991 more rows</span></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true"></a></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true"></a><span class="co"># Add prefixes to families and IDs to recognize them later</span></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true"></a>fam<span class="op">$</span>fam &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&#39;fam&#39;</span>, fam<span class="op">$</span>fam)</span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true"></a>fam<span class="op">$</span>id &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&#39;id&#39;</span>, fam<span class="op">$</span>id)</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true"></a><span class="co"># Sex values are usually 1 and 2</span></span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true"></a>fam<span class="op">$</span>sex &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, n_ind, <span class="dt">replace =</span> <span class="ot">TRUE</span>)</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true"></a><span class="co"># Let&#39;s make phenotypes continuous.</span></span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true"></a><span class="co"># Draw independently from Standard Normal.</span></span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true"></a>fam<span class="op">$</span>pheno &lt;-<span class="st"> </span><span class="kw">rnorm</span>(n_ind)</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true"></a><span class="co"># Let&#39;s leave maternal and paternal IDs as missing</span></span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true"></a></span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true"></a><span class="co"># Inspect again</span></span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true"></a>fam</span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true"></a><span class="co">#&gt; # A tibble: 1,001 × 6</span></span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true"></a><span class="co">#&gt;    fam   id      pat   mat   sex   pheno</span></span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true"></a><span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt;   &lt;dbl&gt;</span></span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true"></a><span class="co">#&gt;  1 fam1  id1       0     0     2  0.0319</span></span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true"></a><span class="co">#&gt;  2 fam2  id2       0     0     1 -0.108 </span></span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true"></a><span class="co">#&gt;  3 fam3  id3       0     0     2 -0.271 </span></span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true"></a><span class="co">#&gt;  4 fam4  id4       0     0     2  0.944 </span></span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true"></a><span class="co">#&gt;  5 fam5  id5       0     0     1  0.208 </span></span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true"></a><span class="co">#&gt;  6 fam6  id6       0     0     2 -0.139 </span></span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true"></a><span class="co">#&gt;  7 fam7  id7       0     0     1 -0.720 </span></span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true"></a><span class="co">#&gt;  8 fam8  id8       0     0     2 -0.207 </span></span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true"></a><span class="co">#&gt;  9 fam9  id9       0     0     1 -0.832 </span></span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true"></a><span class="co">#&gt; 10 fam10 id10      0     0     2  1.01  </span></span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true"></a><span class="co">#&gt; # … with 991 more rows</span></span></code></pre></div>
<p>Lastly, let’s copy the locus and individual IDs as row and column names of the genotype matrix, respectively. Although this step is not required, it is encouraged for consistency (if present, values are checked when writing files).</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true"></a><span class="co"># Add column and row names from bim and fam tables we just created.</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true"></a><span class="kw">rownames</span>(X) &lt;-<span class="st"> </span>bim<span class="op">$</span>id</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true"></a><span class="kw">colnames</span>(X) &lt;-<span class="st"> </span>fam<span class="op">$</span>id</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true"></a><span class="co"># Inspect again the first 10 individuals and loci</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true"></a>X[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true"></a><span class="co">#&gt;      id1 id2 id3 id4 id5 id6 id7 id8 id9 id10</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true"></a><span class="co">#&gt; rs1    0   2   2   2   2   1   0   2   1    1</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true"></a><span class="co">#&gt; rs2    1   1   1   0   1   2   0   1   1    1</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true"></a><span class="co">#&gt; rs3    1   0   0   2   1   0   1   0   2    1</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true"></a><span class="co">#&gt; rs4    1   2   1   0   2   0   0   1  NA    1</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true"></a><span class="co">#&gt; rs5    1   1   1   1  NA   1   1   2   1    0</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true"></a><span class="co">#&gt; rs6   NA   1   0  NA   1  NA   2   2   2    1</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true"></a><span class="co">#&gt; rs7    1   2   2   1   1   1   1   2   1    2</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true"></a><span class="co">#&gt; rs8    1  NA   1   0   1   2   2   1   2    2</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true"></a><span class="co">#&gt; rs9   NA  NA   0   0   2   1   1   0   0    1</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true"></a><span class="co">#&gt; rs10   1   1   2   2   1   0   1   1   2   NA</span></span></code></pre></div>
</div>
<div id="write-genotypes-to-plink-files" class="section level1">
<h1>Write genotypes to Plink files</h1>
<p>Let’s write this random data to a file. This mode is intended for simulated data, generating dummy BIM and FAM files in the process and writing out all three.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true"></a><span class="co"># Will delete at the end of the vignette</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true"></a>file_plink &lt;-<span class="st"> </span><span class="kw">tempfile</span>(<span class="st">&#39;vignette-random-data&#39;</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true"></a><span class="co"># Write genotypes, along with the BIM and FAM files we created.</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true"></a><span class="co"># Omiting them would result in writing the original dummy version of these tables,</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true"></a><span class="co"># before we edited them.</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true"></a>time_write_genio &lt;-<span class="st"> </span><span class="kw">system.time</span>(</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true"></a>    <span class="kw">write_plink</span>(file_plink, X, bim, fam)</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true"></a>)</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true"></a><span class="co">#&gt; Writing: /tmp/RtmpFOguzA/vignette-random-data283201d3206ad.bed</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true"></a><span class="co">#&gt; Writing: /tmp/RtmpFOguzA/vignette-random-data283201d3206ad.bim</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true"></a><span class="co">#&gt; Writing: /tmp/RtmpFOguzA/vignette-random-data283201d3206ad.fam</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true"></a>time_write_genio</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true"></a><span class="co">#&gt;    user  system elapsed </span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true"></a><span class="co">#&gt;   0.225   0.083   0.281</span></span></code></pre></div>
</div>
<div id="read-back-the-genotypes" class="section level1">
<h1>Read back the genotypes</h1>
<p>Here we demonstrate how easy it is to read the data back. To compare to other packages, we shall time loading all this data.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true"></a><span class="co"># Read the data back in memory.</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true"></a><span class="co"># Time this step</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true"></a>time_read_genio &lt;-<span class="st"> </span><span class="kw">system.time</span>(</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true"></a>    data_genio &lt;-<span class="st"> </span><span class="kw">read_plink</span>(file_plink)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true"></a>)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true"></a><span class="co">#&gt; Reading: /tmp/RtmpFOguzA/vignette-random-data283201d3206ad.bim</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true"></a><span class="co">#&gt; Reading: /tmp/RtmpFOguzA/vignette-random-data283201d3206ad.fam</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true"></a><span class="co">#&gt; Reading: /tmp/RtmpFOguzA/vignette-random-data283201d3206ad.bed</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true"></a>time_read_genio</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true"></a><span class="co">#&gt;    user  system elapsed </span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true"></a><span class="co">#&gt;   0.204   0.019   0.223</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true"></a></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true"></a><span class="co"># Inspect the data we just read back</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true"></a></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true"></a><span class="co"># The same random genotypes (first 10 individuals and loci, now with row and column names):</span></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true"></a>data_genio<span class="op">$</span>X[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true"></a><span class="co">#&gt;      id1 id2 id3 id4 id5 id6 id7 id8 id9 id10</span></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true"></a><span class="co">#&gt; rs1    0   2   2   2   2   1   0   2   1    1</span></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true"></a><span class="co">#&gt; rs2    1   1   1   0   1   2   0   1   1    1</span></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true"></a><span class="co">#&gt; rs3    1   0   0   2   1   0   1   0   2    1</span></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true"></a><span class="co">#&gt; rs4    1   2   1   0   2   0   0   1  NA    1</span></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true"></a><span class="co">#&gt; rs5    1   1   1   1  NA   1   1   2   1    0</span></span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true"></a><span class="co">#&gt; rs6   NA   1   0  NA   1  NA   2   2   2    1</span></span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true"></a><span class="co">#&gt; rs7    1   2   2   1   1   1   1   2   1    2</span></span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true"></a><span class="co">#&gt; rs8    1  NA   1   0   1   2   2   1   2    2</span></span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true"></a><span class="co">#&gt; rs9   NA  NA   0   0   2   1   1   0   0    1</span></span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true"></a><span class="co">#&gt; rs10   1   1   2   2   1   0   1   1   2   NA</span></span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true"></a></span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true"></a><span class="co"># The locus annotations</span></span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true"></a>data_genio<span class="op">$</span>bim</span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true"></a><span class="co">#&gt; # A tibble: 10,001 × 6</span></span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true"></a><span class="co">#&gt;    chr   id     posg   pos ref   alt  </span></span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true"></a><span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;int&gt; &lt;chr&gt; &lt;chr&gt;</span></span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true"></a><span class="co">#&gt;  1 chr1  rs1       0  1000 C     A    </span></span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true"></a><span class="co">#&gt;  2 chr1  rs2       0  2000 G     T    </span></span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true"></a><span class="co">#&gt;  3 chr1  rs3       0  3000 C     T    </span></span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true"></a><span class="co">#&gt;  4 chr1  rs4       0  4000 G     A    </span></span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true"></a><span class="co">#&gt;  5 chr1  rs5       0  5000 G     T    </span></span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true"></a><span class="co">#&gt;  6 chr1  rs6       0  6000 G     A    </span></span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true"></a><span class="co">#&gt;  7 chr1  rs7       0  7000 C     A    </span></span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true"></a><span class="co">#&gt;  8 chr1  rs8       0  8000 G     A    </span></span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true"></a><span class="co">#&gt;  9 chr1  rs9       0  9000 G     T    </span></span>
<span id="cb7-43"><a href="#cb7-43" aria-hidden="true"></a><span class="co">#&gt; 10 chr1  rs10      0 10000 C     T    </span></span>
<span id="cb7-44"><a href="#cb7-44" aria-hidden="true"></a><span class="co">#&gt; # … with 9,991 more rows</span></span>
<span id="cb7-45"><a href="#cb7-45" aria-hidden="true"></a></span>
<span id="cb7-46"><a href="#cb7-46" aria-hidden="true"></a><span class="co"># The individual annotations</span></span>
<span id="cb7-47"><a href="#cb7-47" aria-hidden="true"></a>data_genio<span class="op">$</span>fam</span>
<span id="cb7-48"><a href="#cb7-48" aria-hidden="true"></a><span class="co">#&gt; # A tibble: 1,001 × 6</span></span>
<span id="cb7-49"><a href="#cb7-49" aria-hidden="true"></a><span class="co">#&gt;    fam   id    pat   mat     sex   pheno</span></span>
<span id="cb7-50"><a href="#cb7-50" aria-hidden="true"></a><span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;int&gt;   &lt;dbl&gt;</span></span>
<span id="cb7-51"><a href="#cb7-51" aria-hidden="true"></a><span class="co">#&gt;  1 fam1  id1   0     0         2  0.0319</span></span>
<span id="cb7-52"><a href="#cb7-52" aria-hidden="true"></a><span class="co">#&gt;  2 fam2  id2   0     0         1 -0.108 </span></span>
<span id="cb7-53"><a href="#cb7-53" aria-hidden="true"></a><span class="co">#&gt;  3 fam3  id3   0     0         2 -0.271 </span></span>
<span id="cb7-54"><a href="#cb7-54" aria-hidden="true"></a><span class="co">#&gt;  4 fam4  id4   0     0         2  0.944 </span></span>
<span id="cb7-55"><a href="#cb7-55" aria-hidden="true"></a><span class="co">#&gt;  5 fam5  id5   0     0         1  0.208 </span></span>
<span id="cb7-56"><a href="#cb7-56" aria-hidden="true"></a><span class="co">#&gt;  6 fam6  id6   0     0         2 -0.139 </span></span>
<span id="cb7-57"><a href="#cb7-57" aria-hidden="true"></a><span class="co">#&gt;  7 fam7  id7   0     0         1 -0.720 </span></span>
<span id="cb7-58"><a href="#cb7-58" aria-hidden="true"></a><span class="co">#&gt;  8 fam8  id8   0     0         2 -0.207 </span></span>
<span id="cb7-59"><a href="#cb7-59" aria-hidden="true"></a><span class="co">#&gt;  9 fam9  id9   0     0         1 -0.832 </span></span>
<span id="cb7-60"><a href="#cb7-60" aria-hidden="true"></a><span class="co">#&gt; 10 fam10 id10  0     0         2  1.01  </span></span>
<span id="cb7-61"><a href="#cb7-61" aria-hidden="true"></a><span class="co">#&gt; # … with 991 more rows</span></span>
<span id="cb7-62"><a href="#cb7-62" aria-hidden="true"></a></span>
<span id="cb7-63"><a href="#cb7-63" aria-hidden="true"></a><span class="co"># Quickly test that the inputs and outputs are identical.</span></span>
<span id="cb7-64"><a href="#cb7-64" aria-hidden="true"></a><span class="co"># Genotypes have NAs, so we have to compare this way.</span></span>
<span id="cb7-65"><a href="#cb7-65" aria-hidden="true"></a><span class="kw">stopifnot</span>( <span class="kw">all</span>( X <span class="op">==</span><span class="st"> </span>data_genio<span class="op">$</span>X, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>) )</span>
<span id="cb7-66"><a href="#cb7-66" aria-hidden="true"></a><span class="kw">stopifnot</span>( bim <span class="op">==</span><span class="st"> </span>data_genio<span class="op">$</span>bim )</span>
<span id="cb7-67"><a href="#cb7-67" aria-hidden="true"></a><span class="co"># FAM has mixed content (chars, ints, and doubles).</span></span>
<span id="cb7-68"><a href="#cb7-68" aria-hidden="true"></a><span class="co"># First 5 columns should be exact match:</span></span>
<span id="cb7-69"><a href="#cb7-69" aria-hidden="true"></a><span class="kw">stopifnot</span>( fam[,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>] <span class="op">==</span><span class="st"> </span>data_genio<span class="op">$</span>fam[,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>] )</span>
<span id="cb7-70"><a href="#cb7-70" aria-hidden="true"></a><span class="co"># Exact equality may fail for pheno due to precisions, so test this way instead:</span></span>
<span id="cb7-71"><a href="#cb7-71" aria-hidden="true"></a><span class="kw">stopifnot</span>( <span class="kw">max</span>(<span class="kw">abs</span>(fam<span class="op">$</span>pheno <span class="op">-</span><span class="st"> </span>data_genio<span class="op">$</span>fam<span class="op">$</span>pheno)) <span class="op">&lt;</span><span class="st"> </span><span class="fl">1e-4</span> )</span></code></pre></div>
</div>
<div id="memory-estimation" class="section level1">
<h1>Memory estimation</h1>
<p>One drawback of <code>genio</code> and other related approaches is high memory consumption (see more on that at the end of this vignette). It is entirely possible that an attempt to parse a genotype matrix into memory will fail with an “out of memory” error message. Let’s estimate memory usage here.</p>
<p>Genotypes are stored as integers by <code>genio</code>, which in R on a modern machine (64 bit architecture) consumes 4 bytes! So an <span class="math inline">\(n \times m\)</span> matrix takes up at least <span class="math inline">\(4 n m\)</span> bytes (an R matrix contains an additional constant overhead, which does not depend on <span class="math inline">\(n,m\)</span> and is relatively small for large matrices).</p>
<p>To get an idea of how much this is, let’s assume a standard genotyping array, where <span class="math inline">\(m\)</span> is about half a million. In this scenario, we get a simple rule of thumb that every 1000 individuals consume a little less than 2G:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true"></a><span class="co"># Constants</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true"></a>bytes_per_genotype &lt;-<span class="st"> </span><span class="dv">4</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true"></a>bytes_per_gb &lt;-<span class="st"> </span><span class="dv">1024</span> <span class="op">^</span><span class="st"> </span><span class="dv">3</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true"></a><span class="co"># Example data dimensions</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true"></a>num_ind &lt;-<span class="st"> </span><span class="dv">1000</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true"></a>num_loci &lt;-<span class="st"> </span><span class="dv">500000</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true"></a><span class="co"># Gigabytes per 1000 individuals for a typical genotyping array</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true"></a>bytes_per_genotype <span class="op">*</span><span class="st"> </span>num_ind <span class="op">*</span><span class="st"> </span>num_loci <span class="op">/</span><span class="st"> </span>bytes_per_gb</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true"></a><span class="co">#&gt; [1] 1.862645</span></span></code></pre></div>
<p>This is the amount of free memory required just to load the matrix. Several common matrix operations for genotypes consume even more memory, so more free memory will be needed to accomplish anything.</p>
</div>
<div id="comparison-to-other-packages" class="section level1">
<h1>Comparison to other packages</h1>
<p>There are several R packages that provide BED readers, and to a lesser extent some of the other functionality of <code>genio</code>. Here we compare directly to <code>BEDMatrix</code> and <code>snpStats</code>. Each of these packages has different goals so they are optimized for their use cases. The <code>genio</code> parser is optimized to read the entire genotype data into a native R matrix, so that it is easy to inspect and manipulate for R beginners. Here we demonstrate that <code>genio</code> is not only the fastest at this task, but also the easiest to obtain in terms of requiring the least amount of coding.</p>
<div id="bedmatrix" class="section level2">
<h2>BEDMatrix</h2>
<p>The <code>BEDMatrix</code> package allows access to genotypes from a BED file without loading the entire matrix in memory. This is very convenient for large datasets, as long as only small portions of the data are pulled at once. However, there are some disadvantages:</p>
<ul>
<li>The <code>BEDMatrix</code> return object is not a regular R matrix, which confuses some users.</li>
<li>The <code>BEDMatrix</code> package provides no way to access the full annotation tables (BIM and FAM files).
<ul>
<li>From BIM, only the locus ID and reference allele are returned, merged into a vector of strings and placed as column names.</li>
<li>From FAM, only the family and individual IDs are returned, merged into a vector of strings and placed as row names.</li>
</ul></li>
<li><code>BEDMatrix</code> does not have any write functions.</li>
</ul>
<p>Here is an example of that usage and how the data compares. Note in particular that the <code>BEDMatrix</code> representation of the genotype matrix is transposed compared to the <code>genio</code> matrix:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true"></a><span class="kw">library</span>(BEDMatrix)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true"></a><span class="co"># Time it too.</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true"></a><span class="co"># Although the BIM and FAM tables are not returned,</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true"></a><span class="co"># they are partially parsed and kept in memory,</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true"></a><span class="co"># which can take time for extremely large files</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true"></a>time_read_bedmatrix_<span class="dv">1</span> &lt;-<span class="st"> </span><span class="kw">system.time</span>(</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true"></a>    X_BEDMatrix &lt;-<span class="st"> </span><span class="kw">BEDMatrix</span>(file_plink)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true"></a>)</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true"></a><span class="co">#&gt; Extracting number of samples and rownames from vignette-random-data283201d3206ad.fam...</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true"></a><span class="co">#&gt; Extracting number of variants and colnames from vignette-random-data283201d3206ad.bim...</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true"></a></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true"></a><span class="co"># Inspect the first 10 loci and individuals as usual.</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true"></a><span class="co"># Note it is transposed compared to our X.</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true"></a><span class="co"># Also note the column and row names are different from genio&#39;s.</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true"></a>X_BEDMatrix[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true"></a><span class="co">#&gt;            rs1_C rs2_G rs3_C rs4_G rs5_G rs6_G rs7_C rs8_G rs9_G rs10_C</span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true"></a><span class="co">#&gt; fam1_id1       0     1     1     1     1    NA     1     1    NA      1</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true"></a><span class="co">#&gt; fam2_id2       2     1     0     2     1     1     2    NA    NA      1</span></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true"></a><span class="co">#&gt; fam3_id3       2     1     0     1     1     0     2     1     0      2</span></span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true"></a><span class="co">#&gt; fam4_id4       2     0     2     0     1    NA     1     0     0      2</span></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true"></a><span class="co">#&gt; fam5_id5       2     1     1     2    NA     1     1     1     2      1</span></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true"></a><span class="co">#&gt; fam6_id6       1     2     0     0     1    NA     1     2     1      0</span></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true"></a><span class="co">#&gt; fam7_id7       0     0     1     0     1     2     1     2     1      1</span></span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true"></a><span class="co">#&gt; fam8_id8       2     1     0     1     2     2     2     1     0      1</span></span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true"></a><span class="co">#&gt; fam9_id9       1     1     2    NA     1     2     1     2     0      2</span></span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true"></a><span class="co">#&gt; fam10_id10     1     1     1     1     0     1     2     2     1     NA</span></span></code></pre></div>
<p>Therefore, for very large datasets, if it suffices to access the genotype data in small portions and the user is willing to deal with the limitations of the object, and no detailed annotation table information is required, then <code>BEDMatrix</code> is a better solution than the <code>read_bed</code> function provided in the <code>genio</code> package.</p>
<p>To compare the two genotype reading functions on an equal footing, let us assume that the entire genotype matrix is required in memory and stored in a regular R matrix.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true"></a><span class="co"># This turns it into a regular R matrix.</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true"></a><span class="co"># Since most of the reading is actually happening now,</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true"></a><span class="co"># we time this step now.</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true"></a>time_read_bedmatrix_<span class="dv">2</span> &lt;-<span class="st"> </span><span class="kw">system.time</span>(</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true"></a>    X_BEDMatrix_Rmat &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(X_BEDMatrix)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true"></a>)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true"></a>time_read_bedmatrix_<span class="dv">2</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true"></a><span class="co">#&gt;    user  system elapsed </span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true"></a><span class="co">#&gt;   0.060   0.005   0.066</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true"></a><span class="co"># Now we can test that the BEDMatrix agrees with the original matrix we simulated.</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true"></a><span class="co"># Need to transpose first.</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true"></a><span class="kw">stopifnot</span>( <span class="kw">all</span>( X <span class="op">==</span><span class="st"> </span><span class="kw">t</span>(X_BEDMatrix_Rmat), <span class="dt">na.rm =</span> <span class="ot">TRUE</span>) )</span></code></pre></div>
</div>
<div id="snpstats" class="section level2">
<h2>snpStats</h2>
<p>The <code>snpStats</code> package is the most fully-featured alternative to <code>genio</code>. One of its advantages is its memory efficiency, encoding the entire data in less memory than a native R matrix. However, this memory efficiency comes at the cost of making the data harder to access, especially for inexperienced R users. Like <code>genio</code>, and in contrast to the other packages, <code>snpStats</code> also reads the BIM and FAM tables fully, and provides a BED writer, but it is considerably harder to use.</p>
<p>First we illustrate data parsing. The annotation tables are similar but column names are different, and certain missing values (zero in text files) are converted to <code>NA</code> instead.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true"></a><span class="kw">library</span>(snpStats)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true"></a><span class="co">#&gt; Loading required package: survival</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true"></a><span class="co">#&gt; Loading required package: Matrix</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true"></a><span class="co"># Read data, time it.</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true"></a>time_read_snpStats_<span class="dv">1</span> &lt;-<span class="st"> </span><span class="kw">system.time</span>(</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true"></a>    data_snpStats &lt;-<span class="st"> </span><span class="kw">read.plink</span>(file_plink)</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true"></a>)</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true"></a>time_read_snpStats_<span class="dv">1</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true"></a><span class="co">#&gt;    user  system elapsed </span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true"></a><span class="co">#&gt;   0.062   0.002   0.065</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true"></a></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true"></a><span class="co"># Inspect the data</span></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true"></a></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true"></a><span class="co"># Genotypes.</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true"></a><span class="co"># Note data is not visualized this way.</span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true"></a><span class="co"># This matrix is also transposed compared to the genio matrix.</span></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true"></a>data_snpStats<span class="op">$</span>genotypes</span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true"></a><span class="co">#&gt; A SnpMatrix with  1001 rows and  10001 columns</span></span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true"></a><span class="co">#&gt; Row names:  fam1 ... fam1001 </span></span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true"></a><span class="co">#&gt; Col names:  rs1 ... rs10001</span></span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true"></a></span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true"></a><span class="co"># Locus annotations</span></span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true"></a><span class="kw">head</span>( data_snpStats<span class="op">$</span>map )</span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true"></a><span class="co">#&gt;     chromosome snp.name cM position allele.1 allele.2</span></span>
<span id="cb11-26"><a href="#cb11-26" aria-hidden="true"></a><span class="co">#&gt; rs1       chr1      rs1 NA     1000        C        A</span></span>
<span id="cb11-27"><a href="#cb11-27" aria-hidden="true"></a><span class="co">#&gt; rs2       chr1      rs2 NA     2000        G        T</span></span>
<span id="cb11-28"><a href="#cb11-28" aria-hidden="true"></a><span class="co">#&gt; rs3       chr1      rs3 NA     3000        C        T</span></span>
<span id="cb11-29"><a href="#cb11-29" aria-hidden="true"></a><span class="co">#&gt; rs4       chr1      rs4 NA     4000        G        A</span></span>
<span id="cb11-30"><a href="#cb11-30" aria-hidden="true"></a><span class="co">#&gt; rs5       chr1      rs5 NA     5000        G        T</span></span>
<span id="cb11-31"><a href="#cb11-31" aria-hidden="true"></a><span class="co">#&gt; rs6       chr1      rs6 NA     6000        G        A</span></span>
<span id="cb11-32"><a href="#cb11-32" aria-hidden="true"></a></span>
<span id="cb11-33"><a href="#cb11-33" aria-hidden="true"></a><span class="co"># Individual annotations</span></span>
<span id="cb11-34"><a href="#cb11-34" aria-hidden="true"></a><span class="kw">head</span> (data_snpStats<span class="op">$</span>fam )</span>
<span id="cb11-35"><a href="#cb11-35" aria-hidden="true"></a><span class="co">#&gt;      pedigree member father mother sex    affected</span></span>
<span id="cb11-36"><a href="#cb11-36" aria-hidden="true"></a><span class="co">#&gt; fam1     fam1    id1     NA     NA   2  0.03185523</span></span>
<span id="cb11-37"><a href="#cb11-37" aria-hidden="true"></a><span class="co">#&gt; fam2     fam2    id2     NA     NA   1 -0.10766802</span></span>
<span id="cb11-38"><a href="#cb11-38" aria-hidden="true"></a><span class="co">#&gt; fam3     fam3    id3     NA     NA   2 -0.27135791</span></span>
<span id="cb11-39"><a href="#cb11-39" aria-hidden="true"></a><span class="co">#&gt; fam4     fam4    id4     NA     NA   2  0.94377517</span></span>
<span id="cb11-40"><a href="#cb11-40" aria-hidden="true"></a><span class="co">#&gt; fam5     fam5    id5     NA     NA   1  0.20761255</span></span>
<span id="cb11-41"><a href="#cb11-41" aria-hidden="true"></a><span class="co">#&gt; fam6     fam6    id6     NA     NA   2 -0.13882297</span></span></code></pre></div>
<p>As for <code>BEDMatrix</code>, assuming we ultimately desire to convert the entire data into a regular R matrix, an extra step is required:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true"></a><span class="co"># Transpose, then convert to a regular R matrix.</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true"></a><span class="co"># Let&#39;s time this step too.</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true"></a>time_read_snpStats_<span class="dv">2</span> &lt;-<span class="st"> </span><span class="kw">system.time</span>(</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true"></a>    X_snpStats &lt;-<span class="st"> </span><span class="kw">as</span>( <span class="kw">t</span>(data_snpStats<span class="op">$</span>genotypes), <span class="st">&#39;numeric&#39;</span>)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true"></a>)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true"></a>time_read_snpStats_<span class="dv">2</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true"></a><span class="co">#&gt;    user  system elapsed </span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true"></a><span class="co">#&gt;   0.051   0.014   0.065</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true"></a></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true"></a><span class="co"># Now we can visualize the matrix.</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true"></a><span class="co"># First 10 loci and individuals, as before.</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true"></a><span class="co"># Note that, compared to (genio, BEDMatrix), alleles are encoded in reverse,</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true"></a><span class="co"># so 0s and 2s are flipped in this matrix.</span></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true"></a>X_snpStats[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true"></a><span class="co">#&gt;      fam1 fam2 fam3 fam4 fam5 fam6 fam7 fam8 fam9 fam10</span></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true"></a><span class="co">#&gt; rs1     2    0    0    0    0    1    2    0    1     1</span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true"></a><span class="co">#&gt; rs2     1    1    1    2    1    0    2    1    1     1</span></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true"></a><span class="co">#&gt; rs3     1    2    2    0    1    2    1    2    0     1</span></span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true"></a><span class="co">#&gt; rs4     1    0    1    2    0    2    2    1   NA     1</span></span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true"></a><span class="co">#&gt; rs5     1    1    1    1   NA    1    1    0    1     2</span></span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true"></a><span class="co">#&gt; rs6    NA    1    2   NA    1   NA    0    0    0     1</span></span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true"></a><span class="co">#&gt; rs7     1    0    0    1    1    1    1    0    1     0</span></span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true"></a><span class="co">#&gt; rs8     1   NA    1    2    1    0    0    1    0     0</span></span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true"></a><span class="co">#&gt; rs9    NA   NA    2    2    0    1    1    2    2     1</span></span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true"></a><span class="co">#&gt; rs10    1    1    0    0    1    2    1    1    0    NA</span></span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true"></a></span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true"></a><span class="co"># Again verify that the matrices are identical.</span></span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true"></a><span class="co"># (Here 2-X flips back 0s and 2s)</span></span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true"></a><span class="kw">stopifnot</span>( <span class="kw">all</span>( X <span class="op">==</span><span class="st"> </span><span class="dv">2</span> <span class="op">-</span><span class="st"> </span>X_snpStats, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>) )</span></code></pre></div>
<p><code>snpStats</code> is the only package other than <code>genio</code> to provide a BED writer. Here we demonstrate how hard it is to use it to write our data.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true"></a><span class="co"># Let&#39;s write this to another file</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true"></a>file_plink_copy &lt;-<span class="st"> </span><span class="kw">tempfile</span>(<span class="st">&#39;vignette-random-data-copy&#39;</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true"></a><span class="co"># Copy objects to not change originals</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true"></a>X_snpStats &lt;-<span class="st"> </span>X</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true"></a>bim_snpStats &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(bim) <span class="co"># to use rownames</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true"></a>fam_snpStats &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(fam) <span class="co"># ditto</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true"></a></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true"></a><span class="co"># All data requires matching row and/or column names.</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true"></a><span class="co"># These first two were already done above.</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true"></a><span class="co">#rownames(X_snpStats) &lt;- bim$id</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true"></a><span class="co">#colnames(X_snpStats) &lt;- fam$id</span></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true"></a><span class="co"># Row names here are redundant but required.</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true"></a><span class="kw">rownames</span>(bim_snpStats) &lt;-<span class="st"> </span>bim<span class="op">$</span>id</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true"></a><span class="kw">rownames</span>(fam_snpStats) &lt;-<span class="st"> </span>fam<span class="op">$</span>id</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true"></a></span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true"></a><span class="co"># We shall time several required steps in order to write genotypes in a standard R matrix,</span></span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true"></a><span class="co"># and the related annotation tables, to BED.</span></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true"></a>time_write_snpStats &lt;-<span class="st"> </span><span class="kw">system.time</span>({</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true"></a>    <span class="co"># Transpose and convert our genotypes to SnpMatrix object.</span></span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true"></a>    <span class="co"># We flip 0s and 2s before converting</span></span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true"></a>    X_snpStats &lt;-<span class="st"> </span><span class="kw">as</span>(<span class="dv">2</span> <span class="op">-</span><span class="st"> </span><span class="kw">t</span>(X_snpStats), <span class="st">&#39;SnpMatrix&#39;</span>)</span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true"></a>    </span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true"></a>    <span class="co"># This complicated command is required to write the data.</span></span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true"></a>    <span class="co"># Although X, fam, and bim are passed as for genio&#39;s write_plink,</span></span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true"></a>    <span class="co"># here the name of every column must be specified (there are no reasonable defaults).</span></span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true"></a>    <span class="co"># Interestingly, the parameter names of snpStats&#39; write.plink</span></span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true"></a>    <span class="co"># do not agree with read.plink from the same package.</span></span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true"></a>    <span class="kw">write.plink</span>(</span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true"></a>        file_plink_copy,</span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true"></a>        <span class="dt">snps =</span> X_snpStats,</span>
<span id="cb13-32"><a href="#cb13-32" aria-hidden="true"></a>        <span class="dt">subject.data =</span> fam_snpStats,</span>
<span id="cb13-33"><a href="#cb13-33" aria-hidden="true"></a>        <span class="dt">pedigree =</span> fam,</span>
<span id="cb13-34"><a href="#cb13-34" aria-hidden="true"></a>        <span class="dt">id =</span> id,</span>
<span id="cb13-35"><a href="#cb13-35" aria-hidden="true"></a>        <span class="dt">father =</span> pat,</span>
<span id="cb13-36"><a href="#cb13-36" aria-hidden="true"></a>        <span class="dt">mother =</span> mat,</span>
<span id="cb13-37"><a href="#cb13-37" aria-hidden="true"></a>        <span class="dt">sex =</span> sex,</span>
<span id="cb13-38"><a href="#cb13-38" aria-hidden="true"></a>        <span class="dt">phenotype =</span> pheno,</span>
<span id="cb13-39"><a href="#cb13-39" aria-hidden="true"></a>        <span class="dt">snp.data =</span> bim_snpStats,</span>
<span id="cb13-40"><a href="#cb13-40" aria-hidden="true"></a>        <span class="dt">chromosome =</span> chr,</span>
<span id="cb13-41"><a href="#cb13-41" aria-hidden="true"></a>        <span class="dt">genetic.distance =</span> posg,</span>
<span id="cb13-42"><a href="#cb13-42" aria-hidden="true"></a>        <span class="dt">position =</span> pos,</span>
<span id="cb13-43"><a href="#cb13-43" aria-hidden="true"></a>        <span class="dt">allele.1 =</span> ref,</span>
<span id="cb13-44"><a href="#cb13-44" aria-hidden="true"></a>        <span class="dt">allele.2 =</span> alt</span>
<span id="cb13-45"><a href="#cb13-45" aria-hidden="true"></a>    )</span>
<span id="cb13-46"><a href="#cb13-46" aria-hidden="true"></a>})</span>
<span id="cb13-47"><a href="#cb13-47" aria-hidden="true"></a><span class="co">#&gt; Writing FAM file to /tmp/RtmpFOguzA/vignette-random-data-copy283204177de92.fam </span></span>
<span id="cb13-48"><a href="#cb13-48" aria-hidden="true"></a><span class="co">#&gt; Writing extended MAP file to /tmp/RtmpFOguzA/vignette-random-data-copy283204177de92.bim </span></span>
<span id="cb13-49"><a href="#cb13-49" aria-hidden="true"></a><span class="co">#&gt; Writing BED file to /tmp/RtmpFOguzA/vignette-random-data-copy283204177de92.bed (SNP-major mode)</span></span>
<span id="cb13-50"><a href="#cb13-50" aria-hidden="true"></a></span>
<span id="cb13-51"><a href="#cb13-51" aria-hidden="true"></a><span class="co"># remove the new file, no longer need it</span></span>
<span id="cb13-52"><a href="#cb13-52" aria-hidden="true"></a><span class="kw">delete_files_plink</span>(file_plink_copy)</span></code></pre></div>
</div>
<div id="overall-time-comparisons" class="section level2">
<h2>Overall time comparisons</h2>
<p>Note that reading performance varies on different machines depending on the balance between hard drive access and processor speeds (where the relative bottleneck is). That being said, the <code>genio</code> reader is consistently the fastest, if not a close second (see below for tests on your machine), for several reasons. Genotypes are read very efficiently using <code>Rcpp</code> code, and stored directly into an R matrix, which is most efficient if that is the end goal. The annotation tables are also read most efficiently, using the <code>readr</code> package internally.</p>
<p>The <code>BEDMatrix</code> reader is also written in <code>Rcpp</code>, but its main goal is to provide efficient random access to genotypes stored in a file, which makes obtaining an <code>R</code> matrix more awkward, though surprisingly without paying a time penalty for it. The annotation tables are read with <code>data.table</code>, which is <a href="https://csgillespie.github.io/efficientR/input-output.html">actually the fastest</a>, though the difference in performance is small compared to <code>readr</code> for reasonably-sized files. However, <code>BEDMatrix</code> does not process or return full annotation tables, which gives it an unfair advantage compared to <code>genio</code>, as <code>BEDMatrix</code> takes shortcuts to only read the columns it needs. If the annotation tables are needed, reading them will incur an additional (and in some cases considerable) time penalty.</p>
<p>The <code>snpStats</code> reader is written in <code>C</code>, so it is also very fast for its initial step, but converting the genotypes from the <code>snpStats</code> format into a native R matrix proves too expensive. The annotation tables are read with the base function <code>read.table</code>, which is also the slowest. In terms of completeness of output (full genotypes and annotations tables), only this package matches <code>genio</code>, which makes it the fairest comparison.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true"></a><span class="co"># Extract component 3 of each time object,</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true"></a><span class="co"># which is is total time elapsed.</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true"></a><span class="co"># Sum the two steps it takes for each of BEDMatrix and snpStats to obtain a native R matrix.</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true"></a>times_read &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true"></a>    time_read_genio[<span class="dv">3</span>],</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true"></a>    time_read_bedmatrix_<span class="dv">1</span>[<span class="dv">3</span>] <span class="op">+</span><span class="st"> </span>time_read_bedmatrix_<span class="dv">2</span>[<span class="dv">3</span>],</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true"></a>    time_read_snpStats_<span class="dv">1</span>[<span class="dv">3</span>] <span class="op">+</span><span class="st"> </span>time_read_snpStats_<span class="dv">2</span>[<span class="dv">3</span>]</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true"></a>)</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true"></a>names_read &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true"></a>    <span class="st">&#39;genio&#39;</span>,</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true"></a>    <span class="st">&#39;BEDMatrix&#39;</span>,</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true"></a>    <span class="st">&#39;snpStats&#39;</span></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true"></a>)</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true"></a><span class="co"># Create barplot</span></span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true"></a><span class="kw">barplot</span>(</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true"></a>    times_read,</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true"></a>    <span class="dt">names.arg =</span> names_read,</span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true"></a>    <span class="dt">main =</span> <span class="st">&#39;BED reader runtimes&#39;</span>,</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true"></a>    <span class="dt">xlab =</span> <span class="st">&#39;packages&#39;</span>,</span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true"></a>    <span class="dt">ylab =</span> <span class="st">&#39;runtime (s)&#39;</span></span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true"></a>)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAkAAAAGACAMAAAByRC0tAAADAFBMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////isF19AAAACXBIWXMAAA7DAAAOwwHHb6hkAAAZH0lEQVR4nO3deWATZf7H8ac0tEWLIkfl6IogICgqILDgRUBArCyUu2AFlEPAg6ost64Hsgj+2K66eGyXFQUXFREQXBSEyKEI5QYp0OWSsy1CobSl1/f3PJOjaUgyTZ5J08x8Xn+kk/SZYZq+SSZJZ4YRgAQW7BWA0IaAQAoCAikICKQgIJCCgEAKAgIpCAikICCQgoBACgICKQgIpCAgkIKAQAoCAikICKQgIJCCgEAKAgIpCAikICCQgoBACgICKQgIpCAgkIKAQAoCAikICKQgIJCCgEAKAgIpugwo3azoNX6ruLbXbDeAaKsy0S1h7mXfFvmweaK/a3N52bL9RJ+Zzb/6u4TKTJcB7WI2Ya/za5vs11gs0Wr7dOPNPi3SxHr4uzZpjE0mmsXYFn+XUJnpNaD6CQkDW/CCNikBNUxQjFMCapyQ8Egtxupc8mWR0gEtT0g47O8SKjO9BtRTfJ2v/Op4QImOb/GARvIvV/7E2Ku2m943m09tTFjPp5a9EDfqv8ptl+YOfWRocrYy/d9xA5ILbAE5RpTO5Jh+02zOI8ozm9+03vTtM3+afI5o3SjGOkwuVJ7C5pjNGZN6jvyaLs7sM3QZlVki5b8/vMfIf14N6D2jPV0HlMrYDPcB0ZmqrLbtpgmMfRXFFlPeQOW5rU8+0cl6yuQf+O+fpoqpXuEiIKcR9pmcFpDIWA5RjvKv8ZvGh/GhMVk0Q5knX3kKG8lYS3Ht5bvE5d/KLPHC7cpkk4yKva9k6TWgzunph5eaWeSvSkC90xWXSwMiM2PnraP5L7s2i/6cZrKw8QufYGwK0RDGRr/Vg7G/EH3Hn+ye6cJ/szwgpxH2mZwW4BIQ6zbkevEwd/wTxoZYSuwBRTzxELNd1i2zxCQ+498GMTYsOPeZv/QakNUN35PzRvRip4ASHRu1/Jfd/jhRwQ2sP7/2IKuWRw826U2UEcb4y7b7WcRBpageZUbYZnJegEtAr/BnLyZmcNqI5gG9TdSOsTlEbRjLdl7ifew6/pSZ0GFgxd9fMvQdEHu22FNAYxizbnqIX/ZqUn7PXSZPnvxHa1iXN8zjDzvxVFSV9eNXt4mAnEfYZnJegEtAe4gKI1hX14AOkbhMIxrOWJbzEnvz3gd/EmJPYPoNqKPFYvlmuLKlzAPqZlFkOAUUr/wuBf7L5o8x9K2jsxV0qk+EeBOAB5SuPL2ILHqUGWGbyXkBLgGl85uirw0oi2yXIiDnJe6+RXnjocueCr+7pOg1IGUjmpqxGA8b0SWNmanAepPtl72FsT9bO8ssac8ihn55ojoP6Ahjz/NvZouAnEbYZ3JegB8BOS+RCleOqMMTqpdXUXeTJvQd0N3s+hL3AS1n7CHbTbZfdqayzUynUlPzjimDsmxPYXdah/dwHuEuoKHK09YRXwJyWmJuauoZKrbcw9iugN8/WtJrQPxVWPqul/lX51dh6XkioAHp6alzqzG2wTba3sKDrEE6nYhljQr5r/z+i3nDRUDUmbGJF7Y2Ul6FlY5wF9A0xhIO7XvYXUAjPATktMRL4axLDl0ZyMIuBOEe859eA7Ix/eS8Ec02lX6UwR63j7a38LOJhd8WxqquoqJYxqqaTCbxILVMGRyhBFQ6wl1AP4k3fljMNQFlh7HwllfdB+S0xMGMRcVW4S/5K/7+kqHvgO78ljwEVOM9x2hHCz+Jt/Lu/YFPbWzA2G2bG7Own4k+bcjYjaut70SXjnATEM2NYqzF7msCoolhTm8klg3IaYl5STfyychnrlTY3aQJXQaUY90w3Xa2RFzLtjhk0+/K100HikpH/89isW+3Zm49Y524uuco0T6LJUtc+W1fMf1o2VNmhNNMpdM5O07yzRjLAcdNGy27lQVsOU4nLJZLdNBiKSTbZZpy6fRvUsHRn9NyA3B3BJQuA4KKg4BACgICKQgIpCAgkIKAQAoCAikICKQgIJCCgEAKAgIpCAikICCQgoBACgICKQgIpCAgkIKAQAoCAikICKQgIJCCgEAKAgIpCAikICCQgoBACgICKQgIpCAgkIKAQAoCAimBDmjEvUGUEuAfDgIfUNPXPwiawWMD/MNBBQT0ybqgSUJAgYeAQAoCAikICKQgIJDiY0Dp8yaMmfB/3xSUewYEpHM+BXTmsSrhMc1uqc7qfFzeWRCQzvkU0H2NluWIr4enmb4r5ywISOd8CehE6YlkRo8u5zwISOd8CegAs5/kkaYMLuc8CEjnfAmopGH3k8pE8fKb5pdzHgSkcz5tA6XWZS3jhvR/uBZ7tryzICCd8+1lfMFnT/f6o3nQK4fKPQcC0jm8kQhSfAto55Th/1BOiLZgbjnnQEA651NAn5pqN2N3neVTw83lnAUB6ZwvAeXfNOwq7a7TjRAQ2PkS0B52jl/+XGURAgI7XwI6bH0j8Zl65xEQ2Pi0DVRrkPgYPqdpt0tlA9o/2u6hgy6zICCd8ymg/1atPYB/2dvghtgyAZ390K6xxWUWBKRzvr2M3zNpnPhydlqb3u4HdPjZ5QYEpHMav5GIgIzGz4Bee9/97QjIaPwMqK6HV2EIyGj8DOh0pvvbEZDRYBsIpGi8VwYCMhqN98pAQEaj8V4ZCMhoNN4rAwEZjcZ7ZSAgo9F4rwwEZDQa75WBgIxG470yEJDR4I1EkIKAQAoCAikICKQgIJCCgEAKAgIpCAikICCQgoBACgICKQgIpCAgkIKAQAoCAikICKT4EtDVk4X88vD0QS9972kIAjIaXwKysDSib6Ki2zdnT3kYgoCMxteAimLiLhD9EPmF+yEIyGh8DWg/Oywmh3l4CEJARuNrQPvYeTE5tq/7IQjIaHwLKHlbZot3+dTv9V53PwQBGY0vAe1oFsFYuOk8ralZ/6z7IQjIaHx7H6j4hGX+9Gz6IH6vhwEIyGjwRiJIQUAgBQGBFJ8Cys1xyHM/AgEZjU8BtWMOOMwvKHx7BHqL9V5otcb9CARkNGUCupivNvze6e5uTRtjP1tPXddjJyIgnXMEtPnFeyIZq/VYcpa34dPednfrqdKz9axz+RYC0jlbQN+1jRnyzorN6z6fZo4e4zUh7/AUZjTWgCZ2Xllov+X0X1vs9ntxCMhorAEdKHNb9mnV2XC2HrAq+yrs3Mqj5ZsNZ+sBK0dA+aMeoKM3sfAF5ZoNZ+sBK0dAyVGvUVL9/dNqqb6U9wYBGY0joL5DiJpNo1Nsv7fhOFsPlOUIKH4MHWHr6TJL9TwYZ+sBV46AZtb9JqFWYdHrVS94Hoyz9YArR0DZ7VmVj2g0m+p5LM7WA9cofRlftOcY0frNXsbibD1wDWtAiy863/bTHvdjcbYeuIY1oEW3TdxnuyF/SY/OpzwMxtl6wJXtKezsSzVi+0+Y+ZdxD1Rr+x/Po3G2HnDh2AbKWf5893s69p29y9toVQjIaPBH9SAFAVUaR/8XPEdK/F1rBFRZ7DPVC57IDf6uNgKqLHY2Cd49ta71D/6uNgKqLHQQUMavl2XvBQTkt5APKL0tY5YJYz3sclpOCMhvoR7Q/hpdFjeyLK/u4UOuckJAfgv1gMa1LaKWFloVliFzLyAgv4V6QHETSQR0hW2TuRcQkN9CPaAXOykBbbAeRNNfCMhvoR7QvuseT22+4odbekndCwjIb6EeEK1tIQ7b0kdiv2ZCQBJCPiAq2rN01RHJewEB+S3kAypI26XIkbkXEJDfQj2gvfVtxx6zeBuO/cICJtQDGnrzx7+mCbmeB2O/sAAK9YB6T1IfjP3CAijUA5oTrzoW+4UFUqgHVGju+/ESwcNxNwj7hQVWqAe05QbVjWjsFxZIoR7Qk7dtLigSvAzWy35hn9wbRH08rFSoB/ToK+UYrZP9wqbGfxA0b9f2sFKhHtDkzn7/Yb6TEAloRPBW6mu9BvRd/Y5vvC385u+iBASkRrcBDYi1cU3AJwhIjW4DKg+9nK0HAbmqoID0crYeBORKNqBVn9CqFJszngfr5Ww9CMiVbECPNqCeNWw2eRvu/mw9hyfZNXANCwG50GdAXI7tTzQuFnsb7v5sPSdm2TUMibP1ICBXGgRkXmz9Gr3D30UJeApTo8+A8uLja7eNFzqVfmDqDwSkRp8B5Scm3vxAouKf5ZottM/Wg4BcabFnqofXVe6F9tl6EJCrCj+8S2ifrQcBudIgoKJZnTsosBEdULoN6D3W7sUJwv+8DdfHXhkIyJUGAfWKUx+sl70yEJArDQK6/zX1wXrZKwMBudIgoDdaXFIbq5u9MhCQKw0C+v7OeuNne/+DMt3slYGAXFXMH5TpZq8MBOSqgt4H0steGQjIVUW9kaiXvTIQkAsNAprS02avv4sSEJAa3QY0azg3qBnrdczfRQkISI1uA7IqeflGT6crLBcEpEbnAdHViC/8XZSAgNToPaAjYdgvLKB0G9AaZWeLd1pVV31D2hsEpEa3AXWPUjT52t8lKRCQGt0GdOKCv4twhoDU6DaguOf8XYQzBKRGtwGtaHjR32U4QUBqdBvQj4/f/NycZO6kv4sSEJAa3QbUt67NT/4uSkBAanQbkDYQkBoE5BUCUoOAvEJAahCQVwhIjZEDunqykF8enj7ope89DUFAaowckIWlEX0TFd2+OXvKwxAEpMboARXFxF0g+iHSwx99ICA1Rg9oPzssJod5eAhCQGqMHtA+62nBx/Z1PwQBqTF2QMnbMlu8y6d+r/e6+yEISI2RA9rRLIKxcNN5WlOz/ln3QxCQGiMHRFR8wjJ/ejZ9EO9p3x8EpMbYAalCQGoMHhAOMCXL0AHhAFPyDB0QDjAlz8gB4QBTGjByQDjAlAaMHJDHA0wdmWw/W08sztajwsgBeTzA1FHH2XpuWesyCwJyYeiAcIApecYOSBUCUoOAFDhbj78QkAJn6/EXAlLgbD3+QkBeISA1Bg8IH6bKMnRA+DBVnqEDwoep8owcED5M1YCRA8KHqRowckA4W48GjBwQztajAUMHhA9T5Rk7IFUISA0C8goBqUFAXiEgNQjIKwSkBgF5hYDUICCvEJAaBOQVAlKDgLxCQGoQkFcISA0C8goBqUFAXiEgNQjIKwSkBgF5hYDUICCvEJAaBOQVAlKDgLxCQGoMHhD2C5Nl6ICwX5g8QweE/cLkGTkg7BemASMHhP3CNGDkgLBfmAaMHBD2C9OAoQPCfmHyjB2QKgSkxuAB4Y1EWYYOCG8kyjN0QHgjUZ6RA8IbiRowckB4I1EDRg7I4xuJx0tPtuK6Ik0HjgqaBzwG1C54K/WEx4BqBW+lRtUP7huJ6fazPU2KP+Yyyz8nBdF6Dz/HlmCu1DseVurStCCu1GQPB45Xp/EbiWA0Gr+RCEaj8dl6wGg0PlsPGI3GZ+sBo8E2EEjR+MNUMBqNP0wFo9H4w1QwGo0/TAWj0fjDVDAajffKAKPReK8MMBp8mApS8EYiSEFAIAUBgRQEBFIQEEhBQCAFAYEUBARSEBBICY2ALlvyKuYfsnA/Hi4Sk6kWqwtE2/iXDdY/4i2yWIqViRzL5mCuaKURGgGlsvSK+YeYotk2PtnKOs3WErVUJlqk8lsvMmZRRv6LOe9i+saGCl7RstK68vWrPrmQKGvCqdKby1wJkNAI6FzyxYr5h9jErKwz65vewh9lWv0pS1HAA+qTlfXbyjZRW0RAJusHyXEm54Bikyt4Rcu4clvrzbmn/26aQpRe+jdbLlcCpFIFVLw3i7YrfzFyOvUyv8yzFBxdk0bi+UJ5ZjiyrzDAa8BeFZcLxONIq/6OW1smiMvcZt1FQF3rieew7IiuIqDCQ9vOE5VY6jybmmspLNggVnTXTv6NDEuu9mv32y+n+eWZrbTdIh5aDhzKW79D3CNrmXhwpOeq0++LWApf99xde/g/b7tinStQKlNAW+uzsFF38//LxzswFv22+B8021SV3Z9Du8Rv9JfG/GH608CugjWg96rluwmI5rHjPKD3TOLpamHMazygr+owZnqBrjY3xcSnse/q3CpWdFHVvUTdzCVar9tF8Tz1eDGl3D6aVQ17nihxQFt+j3xB9AX7XAw4uqxkY2N260x6pSpjNT4j5Yp9rkCpRAHl1H4i81AblkzFLe87mD2TfckDitlY8C/2vhJQdq3uR7NetP5fCxg23GJZ816dOXyyVSfrRnSJI6BtbB0PaHG35/h0n6dn1KaLUf3SM6az3cpTWBq7Y8VlpfRH7i/5T7XDmq/bC9U3Zs9jSyilSrvjF8ewA5TIuh/5LbHKAcqqFT1+rXXrXTxrrWYzMtPjalqv2OcKlEoU0PvV+dPWJh7QsjBxjI8uXcUjEJ+4dbIS0LuR5/iTxR2JAV0H63Zz7Doq3YjOdwR0ms0XAX1Uv4SuVFvDAzo34zh/smIrbQHNIutD5ZHr5tSdrf26PdaWXyxNoxS2lT/asNWUGJVBdLXBOKJDw2qyqC4fXrU2s3VWkXhUylGu2OcKlEoU0NgH+EVhlWR6tfYSrncsvwPEK5621oDGthaDxrUN6DqIjeisw2NMG0s3oskR0CG2VASUGb6JvqxZOENsA6V/ObOTI6B1toBoDru3SPt1m8/unvkLfzxMMfGF54iA2ombEzuLy+Ltszswc7F1u7l4+4JpTWwB2ecKlEoU0IBHxOX1yTS2ulnoye8A8YRlC2jIQ+LbUxsHdB2s20BFMU+62wZay3aKgOjh8ZTwJPGAriSYOo78hyOgnfaA1rJ+gVi5TSP/wJruopRosgXURdz6/F20c7fy/bl8TUQzR+6p3u35P9sCss8VKJUooD835xfn+VPYXxuJq9nnywY09Q/i1kEPB3QdrAHRnQPcBTSiZq4S0AcN8qqvEgF9fN0B/orHEdAuW0B5TR8N+1b7dTvFn+FTG3dzCqihuLlHL0q8VxlwlH2qNDO8VTbRt7aA7HMFSiUKaC3bTDSbB7SNLSPKqpdUNqAf2Vd88yJqVkDXwRpQWtUZbgJaGjGTlIAywqfccFUENLV2AdES9g0PaK5TQJMa5jzVSPtX8U1H8IuBnZwC4k+ptD38XfqMKUfbmWY6xu+yHXRfb37lWXZZuWKfK1AqUUA08KakYR2j3+NbQ1UTxzdqcKpsQJQYMTIp5q4AvL3ihHVISkoaflODCzygJkmK5TygZklJT3cM63/FGhB1NvFNeR7QhrDubyXeUb3fCbqr40JHQDtMK/jrosmar9ubYX2mdQv73CmgBjVGjLmhdT6VTAlrM2xkm7C/8036KkPXzwgfMcvcjk2/Kq7Y5wqUyhRQ0d/ip2SFL+RTi4f3fJm/wjhlPsivPP0RpZvFG2f/fqLfW4Hth5SNr7iJ4nOvkWYr/p/7Sf6lyzOLxYAcM99U/ty8huhT/v989eA+s/MW9NxOa/q9eMLMIxcr+oJ4mf9Ztyyt161k0ZAeo/mD9KpH+ZU881ZK7Jk6pvdM5R6xvNyn74S9YmreY/8umhefuIgm9soVV+xzBUolCmjL4PNEv4QF7hWn3iT2DPYaUKUKKKN23BefNOsd7NUIHQjIxf7Bd3Z6NT/YaxE6kp4K9hpQ5QoIQhACAikICKQgIJCCgEAKAgIpCAikICCQgoBACgICKQgIpCAgkIKAQAoCAikICKQgIJCCgEAKAgIpCAikICCQgoBACgICKQgIpCCg8luCO+tauE/KDwG5gfvEjSL3R/RCQG7gPnGV0nxBdET7VXwqvU/961t/xSeyEuvFDrugBJTZ/MHc0m+cH1Tr7oVtk4kKJjeL7iSOgbG0zXUNkq4G9yeoUAjIVUpktZlLe1f5gfLr353yRZzpOBU0b7n4nYatSnhAl9q2zi79RnHre778qFYkD6jvzR8uj4vcQdvChy6bEfZmsH+GCoSAXKWwd4lK/mimY11/Fkdm/Zb+VfWkeP7avYTlmW/PoNJvfBV+QhwONZm2sI38ia/FQHo74grRnA+D/TNUIATkKoWJQ0O9W01MZ/40ga2kp8x8suRi4RLWiz1tHWT9xgvikLG5PKC36mRmZWWNaUzfs54rLwVt1YMBAblKMYnLr3lGH8SGNxnEO+lqPcYmfxBq8nz4Hj5h/0Z/cawwik6msdZDStck+rxDmKnH3qCtfMVDQK6sj0DzIotXVkm5Qvm8k8Su4nbLuSXsWN6tnYkc33hWPALl8UegV+4onT9z4e2xATwuc2WDgFylsHn88v6ONDWWxFEzV9Lfq50j2sq2i1dhS9iS0m8sMPGNo694QCvCDxIVPjaFXurDvzU/7Epwf4SKhIBcpVS5cc6KfmGraRF77ZeUxuGv5ube0mrJ1y3aFyvvA3W+Nc/xjaI72yz79x3V3qGSDg1Svu4buZUWsAlr3r+9S7B/hgqEgFylRP/YvnrbFXyzeWq9m+J+fSb6OzrdP6bO4NPWNxL3hL9e+o0zvWt03Bk5n+jSmFujH1zPv/1Wk6jYURnB/QkqFAJypRyGuXwOflggjg8fwIPoVn4IyJUPAaWZpmec6HV3AFem8kNArnwIiD6rz0zmQ4FblxCAgFxdPu7L6MyCQK1HiEBAIAUBgRQEBFIQEEhBQCAFAYEUBARSEBBIQUAgBQGBFAQEUhAQSEFAIAUBgRQEBFIQEEhBQCAFAYEUBARSEBBIQUAg5f8B9obRR8MqtbIAAAAASUVORK5CYII=" style="display: block; margin: auto;" /></p>
<p>Now we compare the writers. Only <code>snpStats</code> and <code>genio</code> have a BED writer. Not only was the <code>genio</code> writer much easier to use, it is also considerably faster.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true"></a>times_write &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true"></a>    time_write_genio[<span class="dv">3</span>],</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true"></a>    time_write_snpStats[<span class="dv">3</span>]</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true"></a>)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true"></a>names_write &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true"></a>    <span class="st">&#39;genio&#39;</span>,</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true"></a>    <span class="st">&#39;snpStats&#39;</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true"></a>)</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true"></a><span class="co"># Create barplot</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true"></a><span class="kw">barplot</span>(</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true"></a>    times_write,</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true"></a>    <span class="dt">names.arg =</span> names_write,</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true"></a>    <span class="dt">main =</span> <span class="st">&#39;BED writer runtimes&#39;</span>,</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true"></a>    <span class="dt">xlab =</span> <span class="st">&#39;packages&#39;</span>,</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true"></a>    <span class="dt">ylab =</span> <span class="st">&#39;runtime (s)&#39;</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true"></a>)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAYAAAAGACAMAAACTGUWNAAADAFBMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////isF19AAAACXBIWXMAAA7DAAAOwwHHb6hkAAAVCklEQVR4nO2dd2AUVf7Av2kkwUQCoRMRD6RLjRhOPBcEpAREBEHIgeVCVQ5IhCCegkgTuIsg9XIinkHkF2kGUeqCYPhBCJJQBCO9hQQhJCE9797M7M5u4mzgZWf3y8bv54/ZN6/NvPfZnTczOwUYgQpgr8AfHRKADAlAhgQgQwKQIQHIkABkSAAyJAAZEoAMCUCGBCBDApAhAciQAGRIADIkABkSgAwJQIYEIEMCkCEByJAAZEgAMiQAGRKADAlAhgQgQwKQIQHIkABk8AS8a5B4buS/7kpzEQYzWxkbJwf6Tdh3rzp6GqYwdnjTpuKKroRSVq4GCTwBA8DEo5f5XHfzHMQwFmIKuk3IK78OH+jF2FCAe2SzjVJWrgYJVAE9hw7t5cc/mCyg71CZ3bKA/kMHt3EHmFd+HX8dOlcPAXI1SKAKSOQfuc3AM0sWkKomcQHSr+K7qlD9lhyRbTBEMZZiMHAhWwyGnbkGw+zsGeHytuOdFgCR2xi7Mn/IS7PSpBpNicwSvmIwLOdzyw2GK3JU+twXX/2aR5jKStWkGwyf7AjrN+sq2z6+37QbUmG1RsaSI154acoJR3QDugAWCg/laQtgbwIsVWJaQwPefwDBfIAAuJANMHwYNJe3HZ7SxiqKba8pffrv5LZMiRKmcCrvZz4XKS2ER/VoJeWdxcxlpWouAzSXZpt94Manj9xhVjWyT9zkLeIiB3QDqoCNqamJ86vACCYL2JUqcYFZBBgB3lIyT5Ri3gCoks86QVOpE2uCVzu55/Z2A9h+NrMWNItZXB8eTlMTJUzhMgKg7httAXxyTWXNAqDNq1KftxkZCLCCWdWY7gWN5k5/DNxTtZtiDw/AIPxSDrMahKVvrlkA7zfT6BgPsIG1BX84UuAN4+VO/Heh9SA8D+A4Y7sAZqqJEqZwGQGBfAvzF7mAZRDmAmrks60A1fPZFoBJ1jVuB5jO2M6QkO36d8MDIKChtCXSFHAd4Cklc5YXTM319P8rrDrCfzhSJ4ZI0RYBQ6FKVFRUhGTMnChhCpcRMJCHZwLsLyNgGJOmQxk7D/CmdY1H+Xq1fy+hwju75YEqYKXRuGtVEARmyAJijRKHmEXAT3KnyHQBQwJ0XQKjV4LHLakTX5FiLQI6mQV2UBMlTOEyAsJ4eOHvBbzJTFNZgFWN7C13KVBrlgMU4A/CsQDLbAzCawDeN0XNAL9oiEqA4L9JX2lTJ1oJ6AW1ZH/Gw2qihClcEQFWNfKYBU9LDqL17wZ8ARsA5msLKORfwt2mqP38mwgb7npUaQHvagkYD968J0sSE3/VEnBW2eyMEBFgVWNq4hG+PfwQYID+3YC9F3RmW0uAvVZ7QalXZAE/pJ6O53HdzLkL/UGy0oZPjb8XcEkaLKNYEd+w/0dLQL47+G1MW+2lJeCSDQFWNU4B+Jaxw258aNadB2AQhtAS61MR3S2nIsD/iJo9FKAeY68BVM0vI2Ay362MZn35NroGQNcCLQHsebm62r8ToJTVFGBV41FfXrYagO8x/bsBX0CV8dLRrpYAt27nLNmj5Q3AUoDerIyA1AbSdzVvgjeXMymHaQo4x7dm3jMifidAKastwFIj29VZWp8ndjigG/AEHJdHuH2n5ZOh7JjRDP+WJcmBxHTr7L8ZjXyQuClPWbHReEqK3GdM5tO8o8ey+UdhyolcZpVYKlxyISmf/Wo05pqjLhmNmeayUjX5RuMvzDTNk6eWGjkZyYeuOaQb6P8AZEgAMiQAGRKADAlAhgQgQwKQIQHIkABkSAAyJAAZEoAMCUCGBCBDApAhAciQAGRIADIkABkSgAwJQIYEIEMCkCEByJAAZEgAMiQAGRKADAlAhgQgQwKQqdwCMv7c0ak89aPwKlZuAadrr3AqIWuEV7GSCwja7VR6koDSkABkSAAyJAAZEoAMCUCGBCBDApAhAciQAGRIADIkABkSgAwJQIYEIEMCkCEByJAAZEgAMiQAGRKADAlAhgQgQwKQIQHIkABkSAAyJAAZEoAMCUCGBCBDApAhAciQAGRIADKVTsCeCMZ+G9ek8evXhReEQmUTMAdassLggLGTGgbdFl4SBpVMwE3PWYVsq/cZxjIf+0B4SRhUMgFHIJOx6I5SMMIBr7V0AJVMQJ7PRsa21S3iwcFjhZeEQSUTwN72nftrvmHs7ZwlHg54ubcDqGwC2Iza4BUA4F49RnhBKFQ6Aaxw+4oPpiz8Kkd4OThUPgEuRuUV8Pxo67m8RDPbHfDaezuovAJChlvPHVAfmOP3bcXqcxCVV4AtQhL0rc9OSAAylU5A7uqXg5sH94v4yVYGEiCKkICEQI+Ql0eN6NcA3ijRzkECRBERUFK/X4YS2hG4SjsLCRBFRMAJuGQOThuinYUEiCIi4AocMQf/NkY7CwkQRWgM6BoUK22DSo5N8tqlnYMEiCIk4MZLHlC1fk1PeOQrGzlIgCiCxwEXv/zn1BnLd9s84UACRKEDMRLgQEgAMiQAGRKADAlAhgQgQwKQcTUBt/PsbTEJEEUVcGByW2+AwL7RGfa0mASIYhLwfXDtYYu3HNj91XSD3xg7FJAAURQBU7rGF5pjrs5tcazCLSYBoigCTpWKy7xa4RaTAFFK7wWlxZ+zr8UkQBRVQF54F3auOniIV2ENCRBFFRDtM5NNrH9ieqBdu6IkQBRVwMBhjDWdzq7ACXtaTAJEUQUMGMPOwh6WBYn2tJgEiKIKmFP3m6GBhUUfeN2yp8UkQBRVQGYncF/FRsE7drWYBIhi2Q0tSj7P2J4D9rWYBIiiCFhX6r7rH5Mr3GISIIoiILbxlOOmiLy4Xl2vVLjFJEAU0yboekRA0KDIOe+P6+Ib/KUdLSYBoqhjQPbmCT3bdh74kc1L/+8LEiAK/SNGAhwICUCGBCBDApBxMQE3TmbZ22ISIIpFQGowgDFybO69y2RssnGPJAmwQ8CJgG7rHjNu9n/l3mWMUGQriQSIogoYF1zEWhvZVrcbtjPPDpPpAcPCZmvnIAGiqAL6TGGSgBw4bDvz+1A1hNMSngoZp52DBIiiCpj8rCxgH9wsJ3d87ZGZtAlyjIDjVYcnNt+yq2H/crOn9X10LwlwiAC2swVwXrzXdYlLH5ryPQlwhABWlLxh69l7lzjZLpAEOEJAwc8/yWTfq0j+/Il0n7D+AlLqg4Lxforl2DheIwGiqAJG1Pns5M8Sd++nWF2D9VzKKDN1tgmvgSNxJQEvTBUpNnmB9dyNVStNNN4rvAaOxJUELNDlcdy0CRJFFVBoGPhZnER6+QWyr5c3SpMAUVQBBx++j0H48MjaPId/x+hMGxlIgCiqgNcaHygokign80poF7Vs7aeLhtfoYOMKUhIgiiqg93v3zJvnM8MUSmv1oXYWEiCKKiCqq81/WcwkgXqeYpaNIZsEiKIK+L5+51kLJS7ZzJvpvt4UKuxhY6eVBIiiChgcZKKcLpzgGR536MzxHz5+stZp7RwkQBShqyIKlzSTd5R8h520kYMEiCJ4WUrJtaRtxlO2X6BBAkRRBGz9nG2NMXHNnhaTAFEUAb0bsNAAE/vtaTEJEMVyeXqB8nnbrpeQkABRVAGGdcqnX5I9LSYBoigCcgcMqBk8QOJZuGBPi0mAKIqAvLCwOl2Uq67+bVeLSYAolivjdujRYhIgCl2e/qAIKJrXNUSGBmEcAZ/Ak5MjJX61p8UkQBRVQP8+erSYBIiiCnh6ph4tJgGiqAJmtbijQ4tJgCiqgO2t6v39o/L/kLkPSIAoQn/I3BsSIAodB5AAB+JKAqaFmkixp8UkQBRVwLxXOUOaQv/z9rSYBIhSZhNU8o9qFX9cFrMWsGO9c9G8t8f1BLD8Kus1890nFgFVn3nWmTSO1FodFxRw1k2n3VDfbU5t+pgIrdVxJQE7vpBY3M7frgNiEiCKKqCnj0yTjRXqeDMkQBRVwEW7HllshgSIYnlWxFsV6vEykABRVAFbHr1dXr77hASIogrYO7zOWwuiOZcr1PMmSIAolhc41DXxY4V63gQJEMVhJ+NIwP1BAvTE4QKOTnt1qXxzwJp/amcgAaIICfivZ82m8MR1HnrVoJ2DBIgiIiCv+sh8dqxWD0YCbOBgAcmQxqcJ7rEkwAYOFvCLcun6+Ho3SYA2jh4DAodIt9FkP97jDgnQxNECtnnVHMw/Uho8HEQCtHD4bmjyVPmBrdend3hBOwMJEIUOxPTEeQJmLi81e+s3E51IgCAVFFD6oX17qpvx3GSOIwH3RwUFXLXxYDPaBIlCY4CeOF5A6rLIMZGLvimwlU4CRBEScK2vu0ftpg39odZnNnKQAFGEBPz5sU3yMyt/me75vXYOEiCKiICLoL5octQo7SwkQBQRAacsz5GYZuNdPyRAFBEBJY/2VC6ZKN5c/VPtLCRAFKExILEutO4zbNBzgfCmjRwkQBSx3dCCtaP7P2UY8t4ZWxlIgCh0IKYnJKAMJMBpkAAFEiAKCdATElAGEuA0SIACCRCFBOgJCSgDCXAaJECBBIhCAvSEBJSBBDgNEqBAAkQhAXpCAspAApwGCVAgAaKQAD0hAWUgAU6DBCiQAFFIgJ6QgDKQAKdBAhRIgCgkQE9IQBlIgNP44wjIvp5dTioJEEVMwOGRtQHAv2N0po0MJEAUIQEroV3UsrWfLhpeo4ONR32TAFGEHtrnM8MUSmv1oXYWEiCKiIAkyDAHZw3QzkICRBERkOlufr9MYY+p2llIgChCY8AEz/C4Q2eO//Dxk7VOa+cgAaIICShc0gwkfIedtJGDBIgieBxQci1pm/FUjs10EiBKBY+Ec3K140mAKLo8M+7CPDMNd5njfF8PdyYh2gKqOXUlwh93moDJC6znUqeaGaC+iXLhVOeyX2s1777r5LX4VbgndT4ZR4ii88k4QhSdT8YRouh8Mo4QReeTcYQoOp+MI0TR+WQcIYrOJ+MIUXQ+GUeIovPJOEIUOhJGhgQgQwKQIQHIkABkSAAyJAAZEoAMCUCGBCBDApAhAciQAGQeWAEpp7DXwDk8sAJCw9AWvaA6AHQ8wkNrV1tFl5rRjQdWwObvsJa8wm3OteyDz9S4yliY9V/fYQ75HxxbwM2U4pvyVYX5R+XL+k6dyd+XcJsHjsuboOxEG+/NtZ+7P6Xk8Y+EG7d2HuWBXGPBmd3XpYQuodI03X0RS+rRxVjE2OWDF3mEMmMqpR+4AgpHu0G9dwJ4aLEPQMgF/jUb3tHDrWosYwP4Jij/7+4Az6c5ZNH/5duZukbGgqbV9IKgFJYKcz0Aev/GWMuWcoZtJ9mQh/2a30lpyzdIz2UpM+ZS+oErYF7AjqylblzAeo9VWUkt2xSzMPc3ss6H1FEEzPaJzd7/SD9HLDnDK/za6Q6tuQCPVYWHqg3hAnxib8fXGMbYDOi84qySS9rqhDRNvLPR/WN5Ri2lH7gC6khXenXnAtq9zgP7YD8LC8xlbDXkyQICJ/LYdZDqgCUfhni+vdvABfThcyNDuIDJPLDU/RorWR7sBo3GpjBFwMJ9PNAyUp5RS+kHqoA02Mmn/whghZ7j4uLiYiGGhUk3HsQrAtJgM5+5ITVad3KbVx25VrrOLEj6DkRKArbyQCrslVJvxoUHVPnWNO6mb/1XGCgC1FL6gSrgBEg31CwI4F39hEEijoVJI6BJwBnYzWcK4HNHLDtr0V+qeE7iAhYykwDpm/4bfJ25M0tKT28crAj4yKPlsPnNFQFqKf1AFZAN6/j0rQBW4ivdWlJ8Pb+UgDz3GCa9RPcHByw66wpf/HQ4YCVgNQ8chKRU+EbOMbKx3OfnZf+dFAFqKf3AHQO6PF/CMuvyMSC0813G5njdKCWAdQ3OZ2x0zSwHLDkW+K7lSdhjJeDJfFYysEFecYv2/AiAXa4Tzvu8P9sORxk75x8hz6il9ANXwOGq3aJaG2ryRgW2eruf50xWWkBitfbTennYen+9XWTUfSRydP32eVYCmrR/+2mPL/kRSJOAfmMG+Te5yX+c9RZeq9Vq1qSGjz9xUJq5Yi6lH8gHYsnhr8TPbMIDV6JCR0g//dnT+CTBUMDenc0DZ98OHad585f9XIjoO+ijO4wN/orPLB/HBSQsGDBKGgfY7U9G9Xp9ldTLqa/1y/l5dN8plxJe/FyeMZfSD1wBkav4pLeN16M7l1RIRFkuroD3/RZ+M9HrMOo6mPhjCiiOfrr14IOoq2DmfKNklOVin4z7w0MCkCEByJAAZEgAMiQAGRKADAlAhgQgQwKQIQHIkABkSAAyJAAZEoAMCUCGBCBDApAhAciQAGRIADIkABkSgExlEhDnio1xxXW2BQlwFkUlmtEkwCnENF/jV6WTfD/Ri/Ufav81D2SE1QsaeUsWkN78mbuWhJtDAtt8ERzNWEFUU79npfsqNnSo2mBiPm4LSuGCArx952x4wX0Xy6vfJmZ9H88LrKB563WLH21XwgXcCW6faUkobt/2/1YFenMBA+us3NzHO4kd9hix6UO32dhtsMIFBcASxkqeMrDz3RMYuwrfsv94XZa2P8fiINfQ7AazJHztcZGx9RDNDkq3ORW1eJktrJLD2IKV2G2wwhUFSLcpLvGVwuk/RkI8e126s7LkdmEc9IfRSiYlYVIwD97lAubXSs/IyBjzJ7YdQuN1vb/CblxQgKc03cg1rAjyaDKE93P3oUpKHDSZ4CFdZG5OGNRbivaLZmPlZ45DDca+CnHz7JWCtvK/xwUFyL+AZd7F8e4xOSyP93NYdynemBYH53MbdWVMTXhT+gXk8l/Aey0t5dO/aBakvReFgisKWManT3dm7wTxz594P3/sm8bYITgi7QXFQZwlYY0nHxy+5gK2eJxmrLDvNBbxIk/61O0Bevy7Cwpwr7Zgy0tu37FYmPn/MX/ymHH3bsN2cRtbdCqWjwO6NspVE4paddi0uqXvYlYS0iBm40DvQ2wNRO5Y3qwbdhuscEEBfns7+Qdv4cPuO/Wq9zk53u97dnVQ7VqvXFUOxJI9PrAkXHshoPNR708ZuzOmkd8ze3jy/CY+QeE3cFtQClcUcN9ZT68sYOycrje2606lFvCz57s3LvZv48CVsZ9KLYCtrQ+ehjOOWxcdcD0BWRdEcqcXOGo9dML1BFQySAAyJAAZEoAMCUCGBCBDApAhAciQAGRIADIkABkSgAwJQIYEIEMCkCEByJAAZEgAMiQAGRKADAlA5n/zsOvQExQ1pQAAAABJRU5ErkJggg==" style="display: block; margin: auto;" /></p>
</div>
<div id="memory-comparison" class="section level2">
<h2>Memory comparison</h2>
<p>High memory consumption is the main sacrifice for the ease of use of native R matrices, and in this respect <code>genio</code> consumes the most memory.</p>
<p>We use the <code>pryr</code> package to compare the memory usage of each native genotype object.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true"></a><span class="kw">library</span>(lobstr)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true"></a><span class="co"># Store directly into a vector</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true"></a>sizes &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true"></a>    <span class="kw">obj_size</span>( X ),</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true"></a>    <span class="kw">obj_size</span>( data_genio<span class="op">$</span>X ),</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true"></a>    <span class="kw">obj_size</span>( X_BEDMatrix ),</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true"></a>    <span class="kw">obj_size</span>( data_snpStats<span class="op">$</span>genotypes )</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true"></a>)</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true"></a>names_sizes &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true"></a>    <span class="st">&#39;original&#39;</span>,</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true"></a>    <span class="st">&#39;genio&#39;</span>,</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true"></a>    <span class="st">&#39;BEDMatrix&#39;</span>,</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true"></a>    <span class="st">&#39;snpStats&#39;</span></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true"></a>)</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true"></a><span class="co"># Create barplot</span></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true"></a><span class="kw">barplot</span>(</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true"></a>    <span class="kw">as.numeric</span>( sizes ),</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true"></a>    <span class="dt">names.arg =</span> names_sizes,</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true"></a>    <span class="dt">main =</span> <span class="st">&#39;Native genotype object sizes&#39;</span>,</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true"></a>    <span class="dt">xlab =</span> <span class="st">&#39;packages&#39;</span>,</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true"></a>    <span class="dt">ylab =</span> <span class="st">&#39;memory (bytes)&#39;</span></span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true"></a>)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAkAAAAGACAMAAAByRC0tAAADAFBMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////isF19AAAACXBIWXMAAA7DAAAOwwHHb6hkAAAgAElEQVR4nO2deUAUZf/AvyyLKIKCiKLimWdpHvEzLV9dyxSVFK+8KLNSibTQCPF47VDLPMpK84hKLX010dC08srNvFJUxDQP8tYU8ATl5vk9z8zu7LDM6szOrss6388fzPF8n2ee2f0w88zsPM8AQRAVgKsrgLg3KBCiChQIUQUKhKgCBUJUgQIhqkCBEFWgQIgqUCBEFSgQogoUCFEFCoSoAgVCVIECIapAgRBVoECIKlAgRBUoEKIKFAhRBQqEqAIFQlSBAiGqQIEQVaBAiCpQIEQVKBCiChQIUQUKhKgCBUJUgQIhqkCBEFW4nUDdDYZjbJpoMCSWSNiflFRESFdDnAM3xpdpB1cMhnnCQl9DtLINOHYnnIvbCeQL8BybzgWYWyJhEEAuIeUhzIEb48u0g7MAMcJCMLRTtgHH7oRzcUeBIInYFOjFQR85cGOOEWjUoMnKNuDYnXAubilQg1xBoN1jekWM2U1nJjYDiP2FO/oPMXRn54X+ht6EXPp4YL+pVy3Z8z/vN/6fzwzPsnkhLcdgmJ7xUZ+X13Ahxf+L6Tn6m3yhzKJwQ9dCujTK0LPoksGwIG1yxAfHSYkCTFhyMoGOvNN7+j9sNX8KEwXf/XJE+Ng9lkrzpL7du1/cUTbHduJrA8+8klktQWUENxTIyxummwX6L3DQf1g9m8ZzR/8xAPu57zCSbK7KVvttNee+25Yu1ugKejpvScump8XH2MJUuvpfA1dk61NCmUMAdhByWw+DSBpA/1p0pc+KEgXwiHLSjXfyp/MBRmI6hYmCTzzKZj2mCBvgmOfBrZ1D+FPYZH7P2IFMlFUUVEZwQ4H846HiRV6g4x7Q+MO4qlAuj/z+DMDm09xn/ztnwjf0VHcrCJokfF4TKpkPExMA2kcFABNIlEYFguBXWwKUzyFkAEDHaWH0DzGXuZ77kpMA1jKBICAqHKDSLVKqcFFOKhD4R/WkxebyAomDOwEMm1gfPHaYN8DI8IJ6H02qD7o0XqC98+bN+9SP7aQoqziojOCOAmXVhCG8QKsbNtxGyGiAo+JGdFEwdCBkKFTMmQHwFyHbAN7nM9MvoG0BSfZgAonSqECB6YR0ZGtS6NGIngAHA2wyl5kXAC0IeR18c5hAuhRCYgDmEevCxTmpQLpkQt4C+JYXSBS8A2A0IRf08Ja4DbQZYBIhW9u122xpRI/jShRlFQeVEdxRILIMYKe5EX0ycSJtFqWUuAqLBv1NUgNeoOvKxcfHvw3mq5rtAF/RybNMIFEaFagvXf0+LZZ8DbCBzu/jzoum7/c1gHOkAbWWCcSuAc8DjCxVuDgnFagznb8EMIoXSBT8BdsM1SBpl1igQ+zkN2UPd1FvEmgNQK30EhUVB5UR3FKg4nbQZg4n0Nd1WZPAWqDfANYcA1hF2poaEtCGz7wA4DfCDllUIFFaNmsuETKbfbPxACfo/E2A4cL3uxVgwSnu4o8KxN3TqQBdSxUuznmWGUbxgy68QKJgevg6b94b0VXYGB1LDppaZBboZCXQM9PE2xEFlRHcUiCyzwNaMIE2ADSdvfNDa4EKq8HIeVA+i4RBkJFjP5+ZHiQ20skrTCBRmlggesI4QOcvgOgMQ8+J4fPBL4cT6EW6Il8HQ0sVLs5JBRpM5wv1MIAXSBRMm8cnzXsjvow/O+tppsdc007coXvINZZLbMcSVEZwT4HIcOA+xZcBLnJtoJICkVFQpw9EEPIGeNN1xcnJ//CZjVwboqghE0iUJhboR4CZdP5bgIWW73cM+DzHhVCB6tJL+i0A75UqXJyTClSrgGu4TOYFEgUvAVhOZ7u1mywWKC2Z2ndlGrBqczvxEkAfLkWUVRxURnBTga5U4gSin38i+bOSWaALZoHoF1wOvuO+wHhSSNs2X/OZ71SHytuvx3BXYaI0sUA3q7Fr7301wOe8uUxCdjNffyKcQBB97XAT8DxeqnBxTnYVFnUttSl4HuUFEgVfrACNztz5COBLywYIiQP4mZD9HjCW34mFAI/c5FJEWcVBZQQ3FYjM4gRaRP9WgPJcy4ZesgTPNQlUUJUaxD5+eiEdVIU2aPNNuT/nGhPluPtAljSxQGQpVyZ/NOHLpP/+tKlVOY9wAlXjinhdonBRzrPmdksUMd0HEgW/R9d7A7TIsWyANqJpzmqVaQGHeYEaAlQMpgwQZxUHlRHcVaC8RkygolfowWT0Yu5CKK2W+UYi4a6berBp7pv0m/IZe0fI/okPwNCRnECWtBICkU20aAj5gUXzZVLGAwzjVgC8PU4P5d8pkirckpMK1L09QPk4FscJJA5eSHWA/hfFG6AHmvZMuBZbiCAQR5cSWUVBZQS3E+gP405uesZoZN/AleS7JNtoTKazuYcOZ5MdxlSWetloPMPHFxw5miPOX5CSSYZAtRJpRUbj33RywWi8xa2/vj/dFM2VSVnHt76ZQLEk+1CeuTCrwoWcuUZjGkk/yMeZfkwVB59Lvl1yA4zM1H3/cjNsJ/7kG87Gw1ZZhaAygtsJpIpT/v4dC8jRStBbYcbh4M/JwAmkjOLr5W3/Gu/+aEsg1mit3KYc+Cv7KWBaK9PtH3sESvEAeF5hHjdCYwIVLetQo3LLmEvKctEWVQf+B680vV7ps14ptMX8k8I8boTGBEIcDQqEqAIFQlSBAiGqQIEQVaBAiCpQIEQVKBCiChQIUQUKhKhCiUBpMX86rR6Im6JEICN4jc2+fxiiJZQJtKFb7YV59w9EtIMygY6TVXVrz77stNogbodSgUjegtq6LnN25dw/HNECigUipPDngb5cv0sEsUcgSv6uVc6oDOJ+2CcQgphQdCOx0Fm1QNwWhXeii45tWLnhwHXn1AVxQxQJlD+rPtfZTR+R4qz6IG6GIoFGeMf8euzy6QNfdaxyzFkVQtwLJQLd8EgyzRV1e8cZlUHcDyUCHYIM8+zUsjO+COJSlAiUV9E0DgC52PRDZ1QGcT8UtYGWQuPRMxfPm9rHt91tZ1UIcS+UXcanjmxSCTyrd1yMT3UgPMqfSMwpOwM8Iq7Hnkdafx7l8Go4iB+fcB7/t8bVe1cmsUeg7w0Or4aDmNZ9odMIf9fVe1cmecgEivzNaQx719V7VyZRLNCPM2YMbDBjxgrxuus/mJmdZRW+aLzzmFzqTUko0ANHsUCfRUY+VT0ycrp43aEBZgJ/tgpv9MIIp+F3xrpyKNADx8GnsHZ7rFY0Wua8rzT4jPXmUaAHDgokFxRIEnsEWtndZhIKpDUc3LUZBdIaigUq2pmQmJduKxUF0hpKBbrRA8BwxesDG8kokNZQKlC7x3YsNxR9qN9oIxkF0hgKBToFKSSRXoSFvyydjgJpDYUCHYRsTqBYG4P3o0BaQ6FAed6LmUB3H50onY4CaQ2lbaC5nq9Et1javqqNETpQIK2h+DJ+VUs9VAi31asHBdIadtxILLhabDMNBdIa9tyJToux+VQrCqQ17BHICDZHWUCBtIYigQaEcARBrZAB0hEokNZQJNCPAT5vxsfHD4G4+MXSESiQ1lB2CrvQuWkynsIQEUrHB5pZcWohCoQIKG5EH2zWPgEFQswovwq7G+2BAiFm7LmM35tk804iCqQ18JFWuaBAkqBAckGBJEGB5IICSYICyQUFkgQFkgsKJAkKJBcUSBIUSC4okCQokFxQIElQILmgQJKgQHJBgSRBgeSCAkmCAskFBZIEBZILCiQJCiQXFEgSFEguKJAkKJBcUCBJUCC5oECSoEByQYEkQYHkggJJggLJBQWSBAWSCwokCQokFxRIkhIC3Sz1+huloEBaQxBo17iW3gCBPedmqikOBdIaJoE2hVYb8vn6Xb+tmmTwjVKhEAqkNXiB4jpvKDCvufxRs8N2F4cCaQ1eoL9LrLtlYwhWGaBAWqPkVdjVDWfsKWRHAzPl11kloUAPOYJAuSM6kDMB4LnUjkIK/jHTGo9AGkMQaG7590lMzaOTAlVdyuMpTGsIAvUdQkjjSeQSHFVTHAqkNQSBIqLIadhOsiBZTXEokNYQBPow+KdBgQWFH3jdUFMcCqQ1BIFutQXdYjISbLyGRyYokNawXMYXpp4lZPsudcWhQFpDfB8o/ViW2uJQIK1hESgtFMAY+3qOquJQIK0hCHTU/5mV9Y3r/AarKg4F0hqCQNGhhaS5kWz0sPlOeDmgQFpDEKhHHGEC3YH9aopDgbSGINC4TpxAO+CamuJQIK0hCPSXz9Dkpuu31emlqjgUSGtYrsK2NgNKH1VPtKJAmkMQKDu/MHXtxtPkps33ocoBBdIagkCGlfzU96Ca4lAgrcELlBMRUTU0gtEJzqkpDgXSGrxAuZGR1TtEcnylqjgUSGtYbiRucURxKJDWEAT65oAjikOBtIYgUG9o8sEp1cWhQFpDEKh495s14MnPrqgrDgXSGuLngYq2j6rq2XXJLRXFoUBao0THwvSvw8Dbu8LbNl/KfF9QIK1hEej0px11Pv1W3L7xiddqu4tDgbSGIFAkVI5ce5ebfeI9u4tDgbSGINDnP+eZZ8/dtrs4FEhriNpAOclrdtzzccTkVnPu1zpCgbSGRaAlwQCgf/UeRx8j1Aw9dO/iUCCtIQj0m67XzswziwL72o41QupHvoPu2XceBdIagkAjnuQeBPoZMmzGGuE4OTNYH77qrs0QFEhrWEbnGMtNssH2SYoJRMjRId5+z0//UzoEBdIalvGBWnJXYavL59mM5QUi5PqCbhUN0iEokNbgBUpLSTkQ2nbZ7k3jfSbbvtAyC0TJPyEdggJpDV4gA1gw2ow91OG+TyuiQFqDF+hSmgVVneNRIK2h8F0ZaV/GRsXO+SnfVjoKpDV4gVbeFK/bnWoj+N+eOs9qjev4QdASGxEokNbgBVr+SNxfphW5iWGdL9kIfqp+Ujabnpqk3yQdgQJpDdMp7Mrb/iH9Yz98N7pDhdD/2Yo9Dynm2ZEjpUNQIK1h6Zm67s2uLdv3nZliO/ZvS5+xCTaGEUKBtIaSRnRx3a4XuZmidQHfSIegQFpD0VVYcjA07zGk/7OBMNpGBAqkNZRdxuevGNXrScPAKSdtBaBAWsOed6YW2L7XiAJpDXsE+t7GL6kEBdIe9nRtRoEQAXu6NqNAiIDirs3TwsIerxIWFided2WRmQbWv+WjQA85irs2b01IeLVJQsJG8bqjI81U/8UqHAV6yLGnazOewhABe7o2o0CIgD1dm7dPsJmEAmkN7NosFxRIEkGg8zJfdZk+LXJyVpKtVhIKpDUsL1sZIyt+f+3KDQ0Zuq42Ov+gQFpDEGh93Zv3ijORH9I/K9FA/vBfIJ2OAmkNQaDfh1YfM2su5eK9wlPhEqECkWEvSKejQFrD0rU52MTue4UfhwxOoOgB0ukokNZQ+Gt8cY2oQirQ6aA50ukokNYo8VPGxfs3g7Z5N+pUe3hAm1zpZBRIa1gEOtyzHEDN2QX3yXD+LUOzbrNt+IMCaQ5BoHMBLeZtWjPa+x1VxaFAWkMQKO4x7neM73V37hF9aMLL87n0pZ9IB6BAWkMQKDyWm+R43OPJxO/0VRtDC/bI0Ms4PhDCIQgUFcZN9oPtgVpzA4blkcNBzxEUCDEjCGTURZ8qvLWuwbO2Y1PhKv27R7ccBULMWK7CvqoCOoAul23HnuK7Nr9R4xoKhJgQ3Qe6tWXxKlsDu/AEDmQDA2U3eu42CoTwWATKP57CkW07+BevquwnjCO1KoWgQAiHINCRmvcdI5G2gsZHs8mVSW16SwegQFpDEOil6kuOHWfYHkVcBiiQ1rB0LBzviOJQIK0hCDQrwhHFoUBaQxCowNB3SSLD9rsyZIACaQ1BoL2VZDSi7wsKpDUEgYY/siu/kKGqOBRIawgCdZ/iiOJQIK0hCBTfudgBxaFAWkMQaFPN9lNnMy6oKQ4F0hqCQANCTFgroAgUSGvYM0biPUCBtIZYoPRjWWqLQ4G0hkWgtFAAY+zrql4XhgJpDkGgo/7PrKxvXOdn4yUYMkGBtIYgUHRoIWluJBs9bD8TLQMUSGtYhneJI0ygO7BfTXEokNYQBBrXiRNoB1xTUxwKpDUEgf7yGZrcdP22Or1UFYcCaQ3LVdjWZuy3+D6ZqopDgbSG6D5QYerajadVFocCaQ1FvTLuDwqkNZT1yrgvKJDWwF4ZckGBJMFeGXJBgSTBXhlycblAq8c7j3i7f3/AXhlycblAXXuMcBo1f7O3VtgrQy6uF2im83aulXqBsFfGfUCBJHFIr4ztAWb0SVZJKJDDKOMCqemVUXzdzP/hEchplHGBsFfGfUCBJMFeGXJBgSTBXhlyQYEkQYHkggJJggLJBQWSBAWSCwokCQokFxRIEhRILiiQJCiQXFAgSVAguaBAkqBAckGBJEGB5IICSYICyQUFkgQFkgsKJAkKJBcUSBIUSC4okCQokFxQIElQILmgQJKgQHJBgSRBgeSCAkmCAskFBZIEBZILCiQJCiQXFEgSFEguKJAkKJBcUCBJUCC5oECSoEByQYEkUShQ2pexUbFzfsq3lY4COY+HQKB/e+o8qzWu4wdBS2xEoEDO4yEQ6Kn6SdwgwKcm6TdJR6BAzsP9BToPKebZkSOlQ1Ag5+H+Av0N58yzE2y8VgwFch7uL1Bx3a4XuZmidQHfSIegQM7D/QUiycHQvMeQ/s8GwmgbESiQ83gIBCL5K0b1etIwcMpJWwEokPN4GAS6LyiQ83goBMIbia7jIRAIbyS6kodAILyR6ErcXyC8kehS3F8gvJHoUtxfILyR6FLcXyC8kehSHgKB8EaiK3kYBOIpyLGZpCmBCr+Y4TwSS322D49A3xtsJmlKoEsVBjuNXnVLfbYokKNxuUCBztva/+qW+mxRIEeDAjmMByjQtLCwx6uEhcWJ150Tztx1tlmFN3rBeW+K9TtTqnKtnLe1J9613tolH+dtbXDdUp98GX/prly2JiS82iQhYaN4XZrw+uiIs1bhXznxXdX/zbWu3EEnbm38fuut5U9x4tYWlvrk1zhxa/F2v6vbwacwRGugQIgq7BFo+wSHVwNxVxz8RCKiNRQLlD4tcnJWUqEz6oK4IUoF2l+7ckNDhq5rnlNqg7gdCgXKD+mflWggf/gvcE51EHdDoUCpcIlQgciwF5xTHcTdUCjQccjgBIoe4JzqIO6GQoGKa0QVUoFOB81xTnUQd0NpI3qbd6NOtYcHtCn1KwKiTRRfxp9/y9Cs22z0B+HBG4mIKhzctRnRGg7u2oxoDQd3bUa0hoO7NiNaw8FdmxGt4eCuzYjWcHDXZkRrOLhrM6I18EYiogoUCFFFGRDoyN+2FgSOHnVqDf74V2UBKUZK6l02m2bkOUPIKdOUkWy8xs/sNN5xRg1cRhkQKDzS1oJA//5OrYFvgsoCDMDwX0pnY4BnMiGjuZmqbC1pBe9xkf8A/GXJt+JbR9WgJLdH6AE8+6XT2ak7ROtLLDiGMiDQul9tLQg4WaD5f90/5p4Yns7MTE8J19PDZ0yVTA56nBldLTPzyu9DYS6NaKVvzkXO1IsFioxwVA1KMigg8eb1dSFPFRMSMle0vsSCY3CRQNnJXF/aY6fI2bS/uLPWtSNF13YSwhb+Ppm3Y89NLu7avhNFxFECnT1N/uG+qFv7rrLJnvQbWw+xB1N2cieQ9ANSJxdZGLiulueAHkdiqgprRwdzkxcDb1OBugC3m0924QS6uPc8/XvwuQ7Gwt0ZZN9tWoOzRvp15xsvK9js9X2naJ4cY/6ZLcfp4r/7iv/cmUVncrxms+Q1cKDYGDQ6mZCCk/vpGdS0wOdyFC4RKO8tHUA3+iUOeuMVjyUR9KxVMMoDakz0J4QtRA59wtPDZzkhN/rTU0CDI44R6OpTAO2H0H/5uy97gO6lHPr/OKGqF4TQ4v3pF/9vFwBdrJ2PGfAC/QV/SAl0DOhJrFVU6w/o/AWPhVSgIy3pfj2bRQZW8m16O3hxKJylNUirsIiQ6VUVdFIfS89Tra+QNJip94Kns0lCo9foPowvIunwOkvPSkrPa6qvFkHWBAHoxxJ+wZTLUbhEoOnll2fvrP08Faj6kLNFzJkZ/luy5nuYBdK9mnW2XXVC4r023D5Qu49jBApreSJjOFCBooN/u5vk9wYVyHNxwb7KA3mBnqu3J3upt51P6hpaGY2/rWjZt4AKVJlvRF8XBCIVpzCBpregs581/oMK1K5x8u0fdZ/xp7Dg4DkZXA0+8r9yusJy+RtdB4tu7vQZTQWq9kf+17CAJEC93dfm6BbT+ni8sDqTj6JnrZvl+6WlT4bD3II5l6NwiUCBMfTPSkgjg6rc4Z2pzpqYXcwCBdLDw7eQS1Z/R9dGhzpEoGNgJORO5QiS7rWMLn5QoYCE9KAzw9pxAh2BJLoQFWJf4Xwj2utLYmlE/2oRqPFLTKCTQE8zHSftpALNZm3ZR2NNAoURXuGCFkO691Cw0VnlqKQ7/mBHILpUL54KtJ7ODHiUNqLj64NH6/dYI5o6c3XaOXqChg3cgjmXo3CFQFdhHeH3aFBHwjlzFbbSmf+aBWInhA1UIHLXuOjNSo4RaJUH6wv5nwhihPmJiYkTqb4hzNpYXqBVcIsu/ABZdhXOGtGZFxboPrI0ovMtAtV8kwlEWk4jV3UHmUAkY+OnkWAWiB6fOIHIXp3feQUbPepTI/YX+iGlsX8NEsoEYg/bJOi5bsOnEyL0Nc+b2s1pqz/sZBLInMtRuEKgk8CGM8qHZWRQb8I5cxTYyGazzAKFE16gLYE1Iyb2dYxA873Z354RVJX/GBhUINbUNAm0WMc+9s2g5Bu0wLeBSN/6Um2gPN2nnEBTW5KF9QkTaKbno0M+bmoWiKbyAmX611HUjD8d38Kj8goqEG0ZcwLpWeN4LVw7a+TS//SK5Zy5M0jf/rX5JoHMuRyFKwTK1bGP62/a5BzELmOpM9mwks6MsRao3mv0W41zjEAbgTUcW0SQPdxDKTl0SSzQZnpEImSBl32d/k0CvREkJdB3sJcT6Dic7PoOE+gssJNoW7NA7BDBCRQZWi2udNk2yaStnAs9KxRYBIIzdGaGH0nwuMVF1H2Vc2aJD70AvGsSyJzLrv2UwCVtoM6h9HQyqmqWIBDp0K2Y3Aq2EijXg/5z5oQ+4RCBrlWeTsg+2oi+U421wLq1KinQzYAxdFstw+wrnBcop3EXCYGO1e1KOIFIi7e99jKBNsMhQs74vU33tJdIoJ/1h5bpU+VvdESjIkK+LJcrEoieK2/U70fOeQ1jJ6ktQJvkIZ+QiVXp+TQRfuIWzLns29HSuESg5MqtJ4R5LiEWgfb7PBPf3FDV6gjUJSju3SZt/FY65Cpsvu6lsY1b0oKW67vEtdVvLSkQWaQLj28RoOD7E2MIiYmJGdXIaxcVqHwMB72eG+0TEzOmq0/zfwgv0Pv6kGIm0PWgx6aOrdOoxV4ypsbsXLNAWXWoUYb28m/R7NC1nTSwYrToFOZVp+e4RwLp4SbRt94LUc959qOFtWj//Q6Prh9HPurX7zy3YMrlKFxzI/H0O+HRO+n0g8n0z2R6aCCpIwZveL8hvzCdDWC1x5BPbkzq+fruqy/GknffdcBG1w2MOtL5NTqzd3SP6CN0OmAV/bOAfpbhbMRH4+vhcefuXYJNYlijqsso9sTvPAPPKHrNzqYvfcE9Rv7aJ4ScMMwi5IjhDDk+qmfchT19lpG04c/f6cuGFKc1WNqdNuCPGX6Xv9Xdr4VFLi8ilwwn6MKoxSTB92rc87HcKJUnZgwNj+bGsdzSbxz5dXCfmTlLww9wC6ZcjqIM/JTBiF1M/3R35nOytwdRY3OqLHLiJlxNgq8LNlpGBHrXd/ZPMV6lxkF1JJ2aJaztXuOuMzfhYrQsUNHcp5sP2OvUTVwb1yZ0pJJfmtyOlY+6YKNlRCDEXUGBEFWgQIgqUCBEFSgQogoUCFEFCoSoAgVCVIECIapAgRBVoECIKlAgRBUoEKIKFAhRBQqEqAIFQlSBAiGqQIEQVaBAiCpQIEQVKBCiChQIUQUKhKgCBZJPIn5YpcHPRD4okAT4mUhQKD1EBgokAX4m1iQ0Xepbri0b2SKtT82KrdfQmczIGiHDbnACZTT9z11LwrWBgY9/HzqXkPz4xr6ddtE1a9v41IrJc+0ePFBQIGsSvCt8uLa3bhvJrfl4wg899OdIftPmKz+v26qYCnQ7tPUtS0JR65arFwd6U4H6Vl+0rof3QbLf86WkaR7TXb0PDxAUyJoE+IKQ4icN5GyXPYRchp/J114X2fnrcCLkGJqkE0vCGs/zbGDOuWQvGyG6sNkLZHa5O4TMepjHkLEGBbImAdgIy19UYPMZu2NhA3mFDWBXfLMgEXrBKD6ITxgbStjgg3PJx0EZmZmZUQ3IZgjfcNtlVXcFKJA1CXr290eq0cIQz4YDqSddBvEpidDwTU82CJ45oX93ttp3LnmdHxu6CiGr2nnow464rPIPHhTIGv4I9KV30QZdwh2SSz2J7MLWG68mwtmcep0JERJGsyNQDj0CTRGNzJPxfZMQB76LoqyDAlmTAGzA+afbk4ls2PoU6slnFa6yAV4PsKuwREi0JCzVX2SvNJlL1nueIKSg5wTydh+a9I2H3W9tcT9QIGsSdJVnre/n8StZDu//mdDA8727d+u0SvyxWdsi7j5Q53o5QkLhY22Svn20wuekuF2thB/7eu8jSyF2y4Imz7h6Hx4gKJA1Cb6/t/ULXU+bzRNrBPQ49obvJnK5f7WgwZf5G4mpnh9YEv7t7d/+kPc3hNyOquf7n+00+eOG5UNGpLt2Dx4oKJA1CoaqPLEon5AzsMuJtSnzoEDWKBDouH5y+vlejzuxMmUfFMgaJYPlrqgJesNJ59XFDUCBrMlSNFx9hp3vOHxoQIEQVaBAiCpQIEQVKBCiChQIUQUKhKgCBUJUgQIhqkCBEFWgQIgqUCBEFSgQogoUCFEFCoSoAgVCVIECITr4CgkAAAAbSURBVKpAgRBVoECIKlAgRBUoEKIKFAhRxf8DW9b4uzPRzXkAAAAASUVORK5CYII=" style="display: block; margin: auto;" /></p>
<p>The <code>BEDMatrix</code> package is ideal for very large files, since data is loaded into memory only as needed. So the main object does not actually hold any genotypes in memory. It is up to the user to decide how much data to access at once to trade-off speed and memory consumption.</p>
<p><code>snpStats</code> manages memory better than <code>genio</code>, but strictly by a factor of 4. Each genotype is encoded as an integer in <code>R</code> (in general) and <code>genio</code> (in particular), which uses 4 bytes (this is actually platform dependent). On the other hand, <code>snpStats</code> uses strictly one byte per genotype.</p>
<p>Overall, there is a narrow window in data sizes in which a genotype matrix is too large for a native R matrix but small enough for a <code>snpStats</code> object. That, and the much more complex interface of <code>snpStats</code>, makes it not very worthwhile in my opinion, unless there is need for functions provided only by that package. I prefer to use <code>BEDMatrix</code> for large datasets.</p>
</div>
</div>
<div id="cleanup" class="section level1">
<h1>Cleanup</h1>
<p>This handy function removes the three BED/BIM/FAM files we generated at the beginning of this vignette.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true"></a><span class="kw">delete_files_plink</span>(file_plink)</span></code></pre></div>
<p>Let’s close by showing the package versions used when this vignette was last built, as the implementations compared here could change in the future.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true"></a><span class="kw">sessionInfo</span>()</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true"></a><span class="co">#&gt; R version 4.0.5 (2021-03-31)</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true"></a><span class="co">#&gt; Platform: x86_64-redhat-linux-gnu (64-bit)</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true"></a><span class="co">#&gt; Running under: Fedora 34 (Workstation Edition)</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true"></a><span class="co">#&gt; </span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true"></a><span class="co">#&gt; Matrix products: default</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true"></a><span class="co">#&gt; BLAS/LAPACK: /usr/lib64/libflexiblas.so.3.0</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true"></a><span class="co">#&gt; </span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true"></a><span class="co">#&gt; locale:</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true"></a><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true"></a><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=C              </span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true"></a><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true"></a><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true"></a><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true"></a><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true"></a><span class="co">#&gt; </span></span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true"></a><span class="co">#&gt; attached base packages:</span></span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true"></a><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true"></a><span class="co">#&gt; </span></span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true"></a><span class="co">#&gt; other attached packages:</span></span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true"></a><span class="co">#&gt; [1] lobstr_1.1.1    snpStats_1.40.0 Matrix_1.3-4    survival_3.2-11</span></span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true"></a><span class="co">#&gt; [5] BEDMatrix_2.0.3 genio_1.0.25   </span></span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true"></a><span class="co">#&gt; </span></span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true"></a><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true"></a><span class="co">#&gt;  [1] Rcpp_1.0.7          highr_0.9           compiler_4.0.5     </span></span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true"></a><span class="co">#&gt;  [4] pillar_1.6.1        pryr_0.1.5          tools_4.0.5        </span></span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true"></a><span class="co">#&gt;  [7] zlibbioc_1.36.0     bit_4.0.4           crochet_2.3.0      </span></span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true"></a><span class="co">#&gt; [10] digest_0.6.27       evaluate_0.14       lifecycle_1.0.0    </span></span>
<span id="cb18-29"><a href="#cb18-29" aria-hidden="true"></a><span class="co">#&gt; [13] tibble_3.1.3        lattice_0.20-44     pkgconfig_2.0.3    </span></span>
<span id="cb18-30"><a href="#cb18-30" aria-hidden="true"></a><span class="co">#&gt; [16] rlang_0.4.11        cli_3.0.1.9000      yaml_2.2.1         </span></span>
<span id="cb18-31"><a href="#cb18-31" aria-hidden="true"></a><span class="co">#&gt; [19] parallel_4.0.5      xfun_0.24           stringr_1.4.0      </span></span>
<span id="cb18-32"><a href="#cb18-32" aria-hidden="true"></a><span class="co">#&gt; [22] knitr_1.33          vctrs_0.3.8         hms_1.1.0          </span></span>
<span id="cb18-33"><a href="#cb18-33" aria-hidden="true"></a><span class="co">#&gt; [25] tidyselect_1.1.1    bit64_4.0.5         grid_4.0.5         </span></span>
<span id="cb18-34"><a href="#cb18-34" aria-hidden="true"></a><span class="co">#&gt; [28] data.table_1.14.0   glue_1.4.2          R6_2.5.0           </span></span>
<span id="cb18-35"><a href="#cb18-35" aria-hidden="true"></a><span class="co">#&gt; [31] fansi_0.5.0         vroom_1.5.3         rmarkdown_2.9      </span></span>
<span id="cb18-36"><a href="#cb18-36" aria-hidden="true"></a><span class="co">#&gt; [34] purrr_0.3.4         readr_2.0.0         tzdb_0.1.2         </span></span>
<span id="cb18-37"><a href="#cb18-37" aria-hidden="true"></a><span class="co">#&gt; [37] magrittr_2.0.1      codetools_0.2-18    ellipsis_0.3.2     </span></span>
<span id="cb18-38"><a href="#cb18-38" aria-hidden="true"></a><span class="co">#&gt; [40] htmltools_0.5.1.1   splines_4.0.5       BiocGenerics_0.36.1</span></span>
<span id="cb18-39"><a href="#cb18-39" aria-hidden="true"></a><span class="co">#&gt; [43] utf8_1.2.2          stringi_1.7.3       crayon_1.4.1</span></span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
